{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  DataBase contruction"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:37:54.868337Z",
     "start_time": "2024-11-14T11:37:54.863923Z"
    }
   },
   "source": [
    "import pandas as pd"
   ],
   "outputs": [],
   "execution_count": 257
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:37:57.308150Z",
     "start_time": "2024-11-14T11:37:54.992251Z"
    }
   },
   "source": [
    "# reading all the data files\n",
    "data_clinical_patient = pd.read_csv('data_clinical_patient.txt', sep=\"\\t\")\n",
    "data_clinical_sample = pd.read_csv('data_clinical_sample.txt', sep=\"\\t\")\n",
    "data_cna_log2 = pd.read_csv('data_cna_log2.txt', sep=\"\\t\")\n",
    "data_cna = pd.read_csv('data_cna.txt', sep=\"\\t\")\n",
    "data_mutations = pd.read_csv('data_mutations.txt', sep=\"\\t\", header=2, dtype={\"Exon_Number\": \"string\"})\n",
    "data_sv = pd.read_csv('data_sv.txt', sep=\"\\t\")"
   ],
   "outputs": [],
   "execution_count": 258
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:37:57.318035Z",
     "start_time": "2024-11-14T11:37:57.314798Z"
    }
   },
   "source": [
    "# removing bad rows\n",
    "data_clinical_sample = data_clinical_sample[4:]"
   ],
   "outputs": [],
   "execution_count": 259
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:37:57.399349Z",
     "start_time": "2024-11-14T11:37:57.392133Z"
    }
   },
   "source": [
    "# matching the sample id to match other tables\n",
    "data_clinical_patient[\"SAMPLE_ID\"] = data_clinical_patient[\"PATIENT_ID\"].apply(lambda x: \"P-\" + x[7:])"
   ],
   "outputs": [],
   "execution_count": 260
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:37:57.557796Z",
     "start_time": "2024-11-14T11:37:57.548511Z"
    }
   },
   "source": [
    "# make all sample id header name the same - \"SAMPLE_ID\"\n",
    "data_clinical_sample.rename(columns={\"Sample Identifier\": 'SAMPLE_ID'}, inplace=True)\n",
    "data_mutations.rename(columns={\"Tumor_Sample_Barcode\": 'SAMPLE_ID'}, inplace=True)\n",
    "data_sv.rename(columns={\"Sample_Id\": 'SAMPLE_ID'}, inplace=True)"
   ],
   "outputs": [],
   "execution_count": 261
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:37:58.169304Z",
     "start_time": "2024-11-14T11:37:57.564999Z"
    }
   },
   "source": [
    "# merge everything\n",
    "merged_clinical_data = data_clinical_patient.merge(data_clinical_sample, on=\"SAMPLE_ID\", how='outer')\n",
    "merged_mutations_data = merged_clinical_data.merge(data_mutations, on=\"SAMPLE_ID\", how='outer')\n",
    "merged_all_data = merged_mutations_data.merge(data_sv, on=\"SAMPLE_ID\", how='outer')"
   ],
   "outputs": [],
   "execution_count": 262
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:38:01.971993Z",
     "start_time": "2024-11-14T11:37:58.248525Z"
    }
   },
   "source": [
    "merged_all_data.to_csv(\"pan_cancer_db_merged.csv\")"
   ],
   "outputs": [],
   "execution_count": 263
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:38:02.089366Z",
     "start_time": "2024-11-14T11:38:02.044378Z"
    }
   },
   "source": [
    "merged_all_data[\"SNP_event\"] = merged_all_data[\"Reference_Allele\"].fillna(\"\").astype(str) + \">\" + merged_all_data[\"Tumor_Seq_Allele2\"].fillna(\"\").astype(str)\n"
   ],
   "outputs": [],
   "execution_count": 264
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:38:02.264737Z",
     "start_time": "2024-11-14T11:38:02.235068Z"
    }
   },
   "source": [
    "data_for_model = merged_all_data[[\"PATIENT_ID\", \"SEX\", \"DIAGNOSIS AGE\", \"SMOKE STATUS\", \"TMB (nonsynonymous)\",\n",
    "                                \"Hugo_Symbol\", \"Chromosome\", \"Start_Position\", \"End_Position\",\n",
    "                                \"Consequence\", \"Variant_Type\", \"SNP_event\", \"Protein_position\", \"Codons\",\n",
    "                                \"Exon_Number\",\"VAR_TYPE_SX\", \"Site1_Hugo_Symbol\", \"Site2_Hugo_Symbol\",\"Event_Info\"]]\n"
   ],
   "outputs": [],
   "execution_count": 265
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:38:03.241864Z",
     "start_time": "2024-11-14T11:38:02.271284Z"
    }
   },
   "source": [
    "data_for_model.to_csv(\"pan_cancer_db_for_model.csv\")"
   ],
   "outputs": [],
   "execution_count": 266
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:38:03.339751Z",
     "start_time": "2024-11-14T11:38:03.333434Z"
    }
   },
   "source": [
    "data_for_model.shape"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(105906, 19)"
      ]
     },
     "execution_count": 267,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 267
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:38:03.437325Z",
     "start_time": "2024-11-14T11:38:03.413247Z"
    }
   },
   "source": [
    "data_for_model.head(20)"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "     PATIENT_ID     SEX  DIAGNOSIS AGE SMOKE STATUS TMB (nonsynonymous)  \\\n",
       "0   Patient0001  Female             67      Unknown         0.333333333   \n",
       "1   Patient0001  Female             67      Unknown         0.333333333   \n",
       "2   Patient0001  Female             67      Unknown         0.333333333   \n",
       "3   Patient0001  Female             67      Unknown         0.333333333   \n",
       "4   Patient0001  Female             67      Unknown         0.333333333   \n",
       "5   Patient0001  Female             67      Unknown         0.333333333   \n",
       "6   Patient0001  Female             67      Unknown         0.333333333   \n",
       "7   Patient0001  Female             67      Unknown         0.333333333   \n",
       "8   Patient0001  Female             67      Unknown         0.333333333   \n",
       "9   Patient0001  Female             67      Unknown         0.333333333   \n",
       "10  Patient0002    Male             75      Unknown                 0.3   \n",
       "11  Patient0002    Male             75      Unknown                 0.3   \n",
       "12  Patient0002    Male             75      Unknown                 0.3   \n",
       "13  Patient0002    Male             75      Unknown                 0.3   \n",
       "14  Patient0002    Male             75      Unknown                 0.3   \n",
       "15  Patient0002    Male             75      Unknown                 0.3   \n",
       "16  Patient0002    Male             75      Unknown                 0.3   \n",
       "17  Patient0002    Male             75      Unknown                 0.3   \n",
       "18  Patient0002    Male             75      Unknown                 0.3   \n",
       "19  Patient0003  Female             45      Unknown                 0.5   \n",
       "\n",
       "   Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "0        KMT2C          7     151836340.0   151836340.0   \n",
       "1         KRAS         12      25398285.0    25398285.0   \n",
       "2         SOX9         17      70119705.0    70119705.0   \n",
       "3          APC          5     112128143.0   112128143.0   \n",
       "4          APC          5     112175147.0   112175147.0   \n",
       "5          QKI          6     163836364.0   163836364.0   \n",
       "6        FBXW7          4     153247289.0   153247289.0   \n",
       "7       PTPN11         12     112910733.0   112910736.0   \n",
       "8         GLI1         12      57863367.0    57863367.0   \n",
       "9          ATM         11     108142138.0   108142138.0   \n",
       "10        SOX9         17      70120165.0    70120165.0   \n",
       "11       EPHB1          3     134920413.0   134920413.0   \n",
       "12       FBXW7          4     153245512.0   153245512.0   \n",
       "13         APC          5     112175609.0   112175609.0   \n",
       "14        RARA         17      38508182.0    38508184.0   \n",
       "15       FBXW7          4     153249384.0   153249384.0   \n",
       "16        KRAS         12      25398284.0    25398284.0   \n",
       "17        TP53         17       7578406.0     7578406.0   \n",
       "18         APC          5     112116592.0   112116592.0   \n",
       "19         APC          5     112175639.0   112175639.0   \n",
       "\n",
       "                             Consequence Variant_Type SNP_event  \\\n",
       "0                       missense_variant          SNP       C>T   \n",
       "1                       missense_variant          SNP       C>A   \n",
       "2                     frameshift_variant          DEL       C>-   \n",
       "3      stop_gained,splice_region_variant          SNP       C>T   \n",
       "4                            stop_gained          SNP       G>T   \n",
       "5                       missense_variant          SNP       G>A   \n",
       "6                       missense_variant          SNP       G>A   \n",
       "7                         intron_variant          DEL    TTTC>-   \n",
       "8                       missense_variant          SNP       G>T   \n",
       "9   splice_region_variant,intron_variant          SNP       G>A   \n",
       "10                    frameshift_variant          DEL       G>-   \n",
       "11                      missense_variant          SNP       G>A   \n",
       "12                      missense_variant          SNP       T>C   \n",
       "13                    frameshift_variant          DEL       C>-   \n",
       "14                      inframe_deletion          DEL     AAG>-   \n",
       "15                      missense_variant          SNP       C>T   \n",
       "16                      missense_variant          SNP       C>A   \n",
       "17                      missense_variant          SNP       C>T   \n",
       "18                           stop_gained          SNP       C>T   \n",
       "19                           stop_gained          SNP       C>T   \n",
       "\n",
       "    Protein_position   Codons Exon_Number         VAR_TYPE_SX  \\\n",
       "0             4822.0  cGt/cAt       57/59  Substitution/Indel   \n",
       "1               12.0  Ggt/Tgt       5-Feb  Substitution/Indel   \n",
       "2              236.0   aCc/ac       3-Mar          Truncation   \n",
       "3              216.0  Cga/Tga      16-Jul          Truncation   \n",
       "4             1286.0  Gaa/Taa       16/16          Truncation   \n",
       "5               47.0  Gaa/Aaa       8-Jan  Substitution/Indel   \n",
       "6              505.0  Cgc/Tgc      12-Oct  Substitution/Indel   \n",
       "7                NaN      NaN        <NA>  Substitution/Indel   \n",
       "8              488.0  Gct/Tct      12-Nov  Substitution/Indel   \n",
       "9             1026.0      NaN        <NA>  Substitution/Indel   \n",
       "10             389.0   ccG/cc       3-Mar          Truncation   \n",
       "11             743.0  cGg/cAg      16-Dec  Substitution/Indel   \n",
       "12             560.0  gAt/gGt      12-Nov  Substitution/Indel   \n",
       "13            1440.0   Cca/ca       16/16          Truncation   \n",
       "14             164.0    AAG/-       9-May  Substitution/Indel   \n",
       "15             465.0  cGt/cAt      12-Sep  Substitution/Indel   \n",
       "16              12.0  gGt/gTt       5-Feb  Substitution/Indel   \n",
       "17             175.0  cGc/cAc      11-May  Substitution/Indel   \n",
       "18             213.0  Cga/Tga      16-Jun          Truncation   \n",
       "19            1450.0  Cga/Tga       16/16          Truncation   \n",
       "\n",
       "   Site1_Hugo_Symbol Site2_Hugo_Symbol        Event_Info  \n",
       "0                NaN               NaN               NaN  \n",
       "1                NaN               NaN               NaN  \n",
       "2                NaN               NaN               NaN  \n",
       "3                NaN               NaN               NaN  \n",
       "4                NaN               NaN               NaN  \n",
       "5                NaN               NaN               NaN  \n",
       "6                NaN               NaN               NaN  \n",
       "7                NaN               NaN               NaN  \n",
       "8                NaN               NaN               NaN  \n",
       "9                NaN               NaN               NaN  \n",
       "10               NaN               NaN               NaN  \n",
       "11               NaN               NaN               NaN  \n",
       "12               NaN               NaN               NaN  \n",
       "13               NaN               NaN               NaN  \n",
       "14               NaN               NaN               NaN  \n",
       "15               NaN               NaN               NaN  \n",
       "16               NaN               NaN               NaN  \n",
       "17               NaN               NaN               NaN  \n",
       "18               NaN               NaN               NaN  \n",
       "19             MAGI2             MAGI2  MAGI2-Intragenic  "
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>SEX</th>\n",
       "      <th>DIAGNOSIS AGE</th>\n",
       "      <th>SMOKE STATUS</th>\n",
       "      <th>TMB (nonsynonymous)</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>SNP_event</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Codons</th>\n",
       "      <th>Exon_Number</th>\n",
       "      <th>VAR_TYPE_SX</th>\n",
       "      <th>Site1_Hugo_Symbol</th>\n",
       "      <th>Site2_Hugo_Symbol</th>\n",
       "      <th>Event_Info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151836340.0</td>\n",
       "      <td>151836340.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>4822.0</td>\n",
       "      <td>cGt/cAt</td>\n",
       "      <td>57/59</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>12</td>\n",
       "      <td>25398285.0</td>\n",
       "      <td>25398285.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>12.0</td>\n",
       "      <td>Ggt/Tgt</td>\n",
       "      <td>5-Feb</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>SOX9</td>\n",
       "      <td>17</td>\n",
       "      <td>70119705.0</td>\n",
       "      <td>70119705.0</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>DEL</td>\n",
       "      <td>C&gt;-</td>\n",
       "      <td>236.0</td>\n",
       "      <td>aCc/ac</td>\n",
       "      <td>3-Mar</td>\n",
       "      <td>Truncation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112128143.0</td>\n",
       "      <td>112128143.0</td>\n",
       "      <td>stop_gained,splice_region_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>216.0</td>\n",
       "      <td>Cga/Tga</td>\n",
       "      <td>16-Jul</td>\n",
       "      <td>Truncation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175147.0</td>\n",
       "      <td>112175147.0</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G&gt;T</td>\n",
       "      <td>1286.0</td>\n",
       "      <td>Gaa/Taa</td>\n",
       "      <td>16/16</td>\n",
       "      <td>Truncation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>QKI</td>\n",
       "      <td>6</td>\n",
       "      <td>163836364.0</td>\n",
       "      <td>163836364.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>47.0</td>\n",
       "      <td>Gaa/Aaa</td>\n",
       "      <td>8-Jan</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>FBXW7</td>\n",
       "      <td>4</td>\n",
       "      <td>153247289.0</td>\n",
       "      <td>153247289.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>505.0</td>\n",
       "      <td>Cgc/Tgc</td>\n",
       "      <td>12-Oct</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>PTPN11</td>\n",
       "      <td>12</td>\n",
       "      <td>112910733.0</td>\n",
       "      <td>112910736.0</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>DEL</td>\n",
       "      <td>TTTC&gt;-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>GLI1</td>\n",
       "      <td>12</td>\n",
       "      <td>57863367.0</td>\n",
       "      <td>57863367.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G&gt;T</td>\n",
       "      <td>488.0</td>\n",
       "      <td>Gct/Tct</td>\n",
       "      <td>12-Nov</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>ATM</td>\n",
       "      <td>11</td>\n",
       "      <td>108142138.0</td>\n",
       "      <td>108142138.0</td>\n",
       "      <td>splice_region_variant,intron_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>1026.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>SOX9</td>\n",
       "      <td>17</td>\n",
       "      <td>70120165.0</td>\n",
       "      <td>70120165.0</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>DEL</td>\n",
       "      <td>G&gt;-</td>\n",
       "      <td>389.0</td>\n",
       "      <td>ccG/cc</td>\n",
       "      <td>3-Mar</td>\n",
       "      <td>Truncation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>EPHB1</td>\n",
       "      <td>3</td>\n",
       "      <td>134920413.0</td>\n",
       "      <td>134920413.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>743.0</td>\n",
       "      <td>cGg/cAg</td>\n",
       "      <td>16-Dec</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>FBXW7</td>\n",
       "      <td>4</td>\n",
       "      <td>153245512.0</td>\n",
       "      <td>153245512.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>560.0</td>\n",
       "      <td>gAt/gGt</td>\n",
       "      <td>12-Nov</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175609.0</td>\n",
       "      <td>112175609.0</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>DEL</td>\n",
       "      <td>C&gt;-</td>\n",
       "      <td>1440.0</td>\n",
       "      <td>Cca/ca</td>\n",
       "      <td>16/16</td>\n",
       "      <td>Truncation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>RARA</td>\n",
       "      <td>17</td>\n",
       "      <td>38508182.0</td>\n",
       "      <td>38508184.0</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>DEL</td>\n",
       "      <td>AAG&gt;-</td>\n",
       "      <td>164.0</td>\n",
       "      <td>AAG/-</td>\n",
       "      <td>9-May</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>FBXW7</td>\n",
       "      <td>4</td>\n",
       "      <td>153249384.0</td>\n",
       "      <td>153249384.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>465.0</td>\n",
       "      <td>cGt/cAt</td>\n",
       "      <td>12-Sep</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>12</td>\n",
       "      <td>25398284.0</td>\n",
       "      <td>25398284.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>12.0</td>\n",
       "      <td>gGt/gTt</td>\n",
       "      <td>5-Feb</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578406.0</td>\n",
       "      <td>7578406.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>175.0</td>\n",
       "      <td>cGc/cAc</td>\n",
       "      <td>11-May</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112116592.0</td>\n",
       "      <td>112116592.0</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>213.0</td>\n",
       "      <td>Cga/Tga</td>\n",
       "      <td>16-Jun</td>\n",
       "      <td>Truncation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Patient0003</td>\n",
       "      <td>Female</td>\n",
       "      <td>45</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.5</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175639.0</td>\n",
       "      <td>112175639.0</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>1450.0</td>\n",
       "      <td>Cga/Tga</td>\n",
       "      <td>16/16</td>\n",
       "      <td>Truncation</td>\n",
       "      <td>MAGI2</td>\n",
       "      <td>MAGI2</td>\n",
       "      <td>MAGI2-Intragenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 268,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 268
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:38:03.566230Z",
     "start_time": "2024-11-14T11:38:03.562791Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# checking if the dates in Exon number are dd/mm or mm/dd by comparing the dates and not dates for gene APC\n",
    "# data_for_model[data_for_model[\"Hugo_Symbol\"].str.contains(\"APC\", na=False)]\n",
    "# result - the format of the dates are total exons-exon number"
   ],
   "outputs": [],
   "execution_count": 269
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:38:03.644067Z",
     "start_time": "2024-11-14T11:38:03.638560Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Function to handle the conversion\n",
    "def convert_exon_number(val):\n",
    "    try:\n",
    "        # First, try to convert to 'Month-Year' format (e.g., 'Sep-89' -> '09/89')\n",
    "        return pd.to_datetime(val, format='%b-%y').strftime('%m/%y')\n",
    "    except ValueError:\n",
    "        pass\n",
    "\n",
    "    try:\n",
    "        # Then, try to convert to 'DD-Mon' format (e.g., '14-Sep' -> '09/14')\n",
    "        date_obj = pd.to_datetime(val, format='%d-%b', errors='raise')\n",
    "        return date_obj.strftime('%m/%d')\n",
    "    except ValueError:\n",
    "        # If neither format matches, return the value as is (non-date-like string)\n",
    "        return val"
   ],
   "outputs": [],
   "execution_count": 270
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:38:21.820662Z",
     "start_time": "2024-11-14T11:38:03.841670Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Apply the function to the column\n",
    "data_for_model['Exon_Number'] = data_for_model['Exon_Number'].apply(convert_exon_number)"
   ],
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/lm/c8j9b0cs7d92s0563h4zcf8w0000gp/T/ipykernel_26383/2571465884.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  data_for_model['Exon_Number'] = data_for_model['Exon_Number'].apply(convert_exon_number)\n"
     ]
    }
   ],
   "execution_count": 271
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:38:21.863857Z",
     "start_time": "2024-11-14T11:38:21.830670Z"
    }
   },
   "cell_type": "code",
   "source": [
    "data_for_model.head(20)"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "     PATIENT_ID     SEX  DIAGNOSIS AGE SMOKE STATUS TMB (nonsynonymous)  \\\n",
       "0   Patient0001  Female             67      Unknown         0.333333333   \n",
       "1   Patient0001  Female             67      Unknown         0.333333333   \n",
       "2   Patient0001  Female             67      Unknown         0.333333333   \n",
       "3   Patient0001  Female             67      Unknown         0.333333333   \n",
       "4   Patient0001  Female             67      Unknown         0.333333333   \n",
       "5   Patient0001  Female             67      Unknown         0.333333333   \n",
       "6   Patient0001  Female             67      Unknown         0.333333333   \n",
       "7   Patient0001  Female             67      Unknown         0.333333333   \n",
       "8   Patient0001  Female             67      Unknown         0.333333333   \n",
       "9   Patient0001  Female             67      Unknown         0.333333333   \n",
       "10  Patient0002    Male             75      Unknown                 0.3   \n",
       "11  Patient0002    Male             75      Unknown                 0.3   \n",
       "12  Patient0002    Male             75      Unknown                 0.3   \n",
       "13  Patient0002    Male             75      Unknown                 0.3   \n",
       "14  Patient0002    Male             75      Unknown                 0.3   \n",
       "15  Patient0002    Male             75      Unknown                 0.3   \n",
       "16  Patient0002    Male             75      Unknown                 0.3   \n",
       "17  Patient0002    Male             75      Unknown                 0.3   \n",
       "18  Patient0002    Male             75      Unknown                 0.3   \n",
       "19  Patient0003  Female             45      Unknown                 0.5   \n",
       "\n",
       "   Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "0        KMT2C          7     151836340.0   151836340.0   \n",
       "1         KRAS         12      25398285.0    25398285.0   \n",
       "2         SOX9         17      70119705.0    70119705.0   \n",
       "3          APC          5     112128143.0   112128143.0   \n",
       "4          APC          5     112175147.0   112175147.0   \n",
       "5          QKI          6     163836364.0   163836364.0   \n",
       "6        FBXW7          4     153247289.0   153247289.0   \n",
       "7       PTPN11         12     112910733.0   112910736.0   \n",
       "8         GLI1         12      57863367.0    57863367.0   \n",
       "9          ATM         11     108142138.0   108142138.0   \n",
       "10        SOX9         17      70120165.0    70120165.0   \n",
       "11       EPHB1          3     134920413.0   134920413.0   \n",
       "12       FBXW7          4     153245512.0   153245512.0   \n",
       "13         APC          5     112175609.0   112175609.0   \n",
       "14        RARA         17      38508182.0    38508184.0   \n",
       "15       FBXW7          4     153249384.0   153249384.0   \n",
       "16        KRAS         12      25398284.0    25398284.0   \n",
       "17        TP53         17       7578406.0     7578406.0   \n",
       "18         APC          5     112116592.0   112116592.0   \n",
       "19         APC          5     112175639.0   112175639.0   \n",
       "\n",
       "                             Consequence Variant_Type SNP_event  \\\n",
       "0                       missense_variant          SNP       C>T   \n",
       "1                       missense_variant          SNP       C>A   \n",
       "2                     frameshift_variant          DEL       C>-   \n",
       "3      stop_gained,splice_region_variant          SNP       C>T   \n",
       "4                            stop_gained          SNP       G>T   \n",
       "5                       missense_variant          SNP       G>A   \n",
       "6                       missense_variant          SNP       G>A   \n",
       "7                         intron_variant          DEL    TTTC>-   \n",
       "8                       missense_variant          SNP       G>T   \n",
       "9   splice_region_variant,intron_variant          SNP       G>A   \n",
       "10                    frameshift_variant          DEL       G>-   \n",
       "11                      missense_variant          SNP       G>A   \n",
       "12                      missense_variant          SNP       T>C   \n",
       "13                    frameshift_variant          DEL       C>-   \n",
       "14                      inframe_deletion          DEL     AAG>-   \n",
       "15                      missense_variant          SNP       C>T   \n",
       "16                      missense_variant          SNP       C>A   \n",
       "17                      missense_variant          SNP       C>T   \n",
       "18                           stop_gained          SNP       C>T   \n",
       "19                           stop_gained          SNP       C>T   \n",
       "\n",
       "    Protein_position   Codons Exon_Number         VAR_TYPE_SX  \\\n",
       "0             4822.0  cGt/cAt       57/59  Substitution/Indel   \n",
       "1               12.0  Ggt/Tgt       02/05  Substitution/Indel   \n",
       "2              236.0   aCc/ac       03/03          Truncation   \n",
       "3              216.0  Cga/Tga       07/16          Truncation   \n",
       "4             1286.0  Gaa/Taa       16/16          Truncation   \n",
       "5               47.0  Gaa/Aaa       01/08  Substitution/Indel   \n",
       "6              505.0  Cgc/Tgc       10/12  Substitution/Indel   \n",
       "7                NaN      NaN        <NA>  Substitution/Indel   \n",
       "8              488.0  Gct/Tct       11/12  Substitution/Indel   \n",
       "9             1026.0      NaN        <NA>  Substitution/Indel   \n",
       "10             389.0   ccG/cc       03/03          Truncation   \n",
       "11             743.0  cGg/cAg       12/16  Substitution/Indel   \n",
       "12             560.0  gAt/gGt       11/12  Substitution/Indel   \n",
       "13            1440.0   Cca/ca       16/16          Truncation   \n",
       "14             164.0    AAG/-       05/09  Substitution/Indel   \n",
       "15             465.0  cGt/cAt       09/12  Substitution/Indel   \n",
       "16              12.0  gGt/gTt       02/05  Substitution/Indel   \n",
       "17             175.0  cGc/cAc       05/11  Substitution/Indel   \n",
       "18             213.0  Cga/Tga       06/16          Truncation   \n",
       "19            1450.0  Cga/Tga       16/16          Truncation   \n",
       "\n",
       "   Site1_Hugo_Symbol Site2_Hugo_Symbol        Event_Info  \n",
       "0                NaN               NaN               NaN  \n",
       "1                NaN               NaN               NaN  \n",
       "2                NaN               NaN               NaN  \n",
       "3                NaN               NaN               NaN  \n",
       "4                NaN               NaN               NaN  \n",
       "5                NaN               NaN               NaN  \n",
       "6                NaN               NaN               NaN  \n",
       "7                NaN               NaN               NaN  \n",
       "8                NaN               NaN               NaN  \n",
       "9                NaN               NaN               NaN  \n",
       "10               NaN               NaN               NaN  \n",
       "11               NaN               NaN               NaN  \n",
       "12               NaN               NaN               NaN  \n",
       "13               NaN               NaN               NaN  \n",
       "14               NaN               NaN               NaN  \n",
       "15               NaN               NaN               NaN  \n",
       "16               NaN               NaN               NaN  \n",
       "17               NaN               NaN               NaN  \n",
       "18               NaN               NaN               NaN  \n",
       "19             MAGI2             MAGI2  MAGI2-Intragenic  "
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>SEX</th>\n",
       "      <th>DIAGNOSIS AGE</th>\n",
       "      <th>SMOKE STATUS</th>\n",
       "      <th>TMB (nonsynonymous)</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>SNP_event</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Codons</th>\n",
       "      <th>Exon_Number</th>\n",
       "      <th>VAR_TYPE_SX</th>\n",
       "      <th>Site1_Hugo_Symbol</th>\n",
       "      <th>Site2_Hugo_Symbol</th>\n",
       "      <th>Event_Info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151836340.0</td>\n",
       "      <td>151836340.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>4822.0</td>\n",
       "      <td>cGt/cAt</td>\n",
       "      <td>57/59</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>12</td>\n",
       "      <td>25398285.0</td>\n",
       "      <td>25398285.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>12.0</td>\n",
       "      <td>Ggt/Tgt</td>\n",
       "      <td>02/05</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>SOX9</td>\n",
       "      <td>17</td>\n",
       "      <td>70119705.0</td>\n",
       "      <td>70119705.0</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>DEL</td>\n",
       "      <td>C&gt;-</td>\n",
       "      <td>236.0</td>\n",
       "      <td>aCc/ac</td>\n",
       "      <td>03/03</td>\n",
       "      <td>Truncation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112128143.0</td>\n",
       "      <td>112128143.0</td>\n",
       "      <td>stop_gained,splice_region_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>216.0</td>\n",
       "      <td>Cga/Tga</td>\n",
       "      <td>07/16</td>\n",
       "      <td>Truncation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175147.0</td>\n",
       "      <td>112175147.0</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G&gt;T</td>\n",
       "      <td>1286.0</td>\n",
       "      <td>Gaa/Taa</td>\n",
       "      <td>16/16</td>\n",
       "      <td>Truncation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>QKI</td>\n",
       "      <td>6</td>\n",
       "      <td>163836364.0</td>\n",
       "      <td>163836364.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>47.0</td>\n",
       "      <td>Gaa/Aaa</td>\n",
       "      <td>01/08</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>FBXW7</td>\n",
       "      <td>4</td>\n",
       "      <td>153247289.0</td>\n",
       "      <td>153247289.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>505.0</td>\n",
       "      <td>Cgc/Tgc</td>\n",
       "      <td>10/12</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>PTPN11</td>\n",
       "      <td>12</td>\n",
       "      <td>112910733.0</td>\n",
       "      <td>112910736.0</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>DEL</td>\n",
       "      <td>TTTC&gt;-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>GLI1</td>\n",
       "      <td>12</td>\n",
       "      <td>57863367.0</td>\n",
       "      <td>57863367.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G&gt;T</td>\n",
       "      <td>488.0</td>\n",
       "      <td>Gct/Tct</td>\n",
       "      <td>11/12</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>ATM</td>\n",
       "      <td>11</td>\n",
       "      <td>108142138.0</td>\n",
       "      <td>108142138.0</td>\n",
       "      <td>splice_region_variant,intron_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>1026.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>SOX9</td>\n",
       "      <td>17</td>\n",
       "      <td>70120165.0</td>\n",
       "      <td>70120165.0</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>DEL</td>\n",
       "      <td>G&gt;-</td>\n",
       "      <td>389.0</td>\n",
       "      <td>ccG/cc</td>\n",
       "      <td>03/03</td>\n",
       "      <td>Truncation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>EPHB1</td>\n",
       "      <td>3</td>\n",
       "      <td>134920413.0</td>\n",
       "      <td>134920413.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>743.0</td>\n",
       "      <td>cGg/cAg</td>\n",
       "      <td>12/16</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>FBXW7</td>\n",
       "      <td>4</td>\n",
       "      <td>153245512.0</td>\n",
       "      <td>153245512.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>560.0</td>\n",
       "      <td>gAt/gGt</td>\n",
       "      <td>11/12</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175609.0</td>\n",
       "      <td>112175609.0</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>DEL</td>\n",
       "      <td>C&gt;-</td>\n",
       "      <td>1440.0</td>\n",
       "      <td>Cca/ca</td>\n",
       "      <td>16/16</td>\n",
       "      <td>Truncation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>RARA</td>\n",
       "      <td>17</td>\n",
       "      <td>38508182.0</td>\n",
       "      <td>38508184.0</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>DEL</td>\n",
       "      <td>AAG&gt;-</td>\n",
       "      <td>164.0</td>\n",
       "      <td>AAG/-</td>\n",
       "      <td>05/09</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>FBXW7</td>\n",
       "      <td>4</td>\n",
       "      <td>153249384.0</td>\n",
       "      <td>153249384.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>465.0</td>\n",
       "      <td>cGt/cAt</td>\n",
       "      <td>09/12</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>12</td>\n",
       "      <td>25398284.0</td>\n",
       "      <td>25398284.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>12.0</td>\n",
       "      <td>gGt/gTt</td>\n",
       "      <td>02/05</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578406.0</td>\n",
       "      <td>7578406.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>175.0</td>\n",
       "      <td>cGc/cAc</td>\n",
       "      <td>05/11</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112116592.0</td>\n",
       "      <td>112116592.0</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>213.0</td>\n",
       "      <td>Cga/Tga</td>\n",
       "      <td>06/16</td>\n",
       "      <td>Truncation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Patient0003</td>\n",
       "      <td>Female</td>\n",
       "      <td>45</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.5</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175639.0</td>\n",
       "      <td>112175639.0</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>1450.0</td>\n",
       "      <td>Cga/Tga</td>\n",
       "      <td>16/16</td>\n",
       "      <td>Truncation</td>\n",
       "      <td>MAGI2</td>\n",
       "      <td>MAGI2</td>\n",
       "      <td>MAGI2-Intragenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 272,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 272
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:38:22.015821Z",
     "start_time": "2024-11-14T11:38:21.996532Z"
    }
   },
   "cell_type": "code",
   "source": [
    "data_for_model[\"Hugo_Symbol\"].value_counts()"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Hugo_Symbol\n",
       "TP53            6719\n",
       "LRP1B           2138\n",
       "APC             1882\n",
       "KRAS            1781\n",
       "EGFR            1402\n",
       "                ... \n",
       "STK24-AS1          1\n",
       "EGFR-AS1           1\n",
       "DPYD-AS1           1\n",
       "MIR4466            1\n",
       "RP11-770J1.3       1\n",
       "Name: count, Length: 479, dtype: int64"
      ]
     },
     "execution_count": 273,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 273
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:38:22.196877Z",
     "start_time": "2024-11-14T11:38:22.177973Z"
    }
   },
   "cell_type": "code",
   "source": [
    "data_for_model[\"Chromosome\"].value_counts()"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Chromosome\n",
       "17    11798\n",
       "2      8468\n",
       "1      7888\n",
       "7      7764\n",
       "3      7694\n",
       "12     7669\n",
       "5      6356\n",
       "4      5822\n",
       "8      4713\n",
       "19     4491\n",
       "X      4341\n",
       "11     4282\n",
       "9      3657\n",
       "16     3449\n",
       "6      3194\n",
       "13     2753\n",
       "10     2587\n",
       "20     1711\n",
       "22     1663\n",
       "14     1648\n",
       "15     1583\n",
       "18     1516\n",
       "21      446\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 274,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 274
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:38:22.412442Z",
     "start_time": "2024-11-14T11:38:22.393755Z"
    }
   },
   "cell_type": "code",
   "source": [
    "data_for_model[\"Consequence\"].value_counts()\n"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Consequence\n",
       "missense_variant                                                                      65649\n",
       "frameshift_variant                                                                    10367\n",
       "stop_gained                                                                            8673\n",
       "intron_variant                                                                         5213\n",
       "splice_region_variant,intron_variant                                                   5136\n",
       "missense_variant,splice_region_variant                                                 1916\n",
       "splice_acceptor_variant                                                                1725\n",
       "inframe_deletion                                                                       1415\n",
       "splice_donor_variant                                                                   1354\n",
       "upstream_gene_variant                                                                  1214\n",
       "splice_region_variant,synonymous_variant                                                471\n",
       "stop_gained,splice_region_variant                                                       322\n",
       "inframe_insertion                                                                       309\n",
       "5_prime_UTR_variant                                                                     267\n",
       "frameshift_variant,splice_region_variant                                                209\n",
       "synonymous_variant                                                                      202\n",
       "protein_altering_variant                                                                126\n",
       "splice_donor_variant,coding_sequence_variant,intron_variant                             115\n",
       "splice_acceptor_variant,coding_sequence_variant,intron_variant                          107\n",
       "stop_gained,frameshift_variant                                                          101\n",
       "start_lost                                                                               96\n",
       "intron_variant,non_coding_transcript_variant                                             74\n",
       "downstream_gene_variant                                                                  48\n",
       "stop_lost                                                                                48\n",
       "splice_donor_variant,coding_sequence_variant                                             41\n",
       "splice_donor_variant,intron_variant                                                      39\n",
       "splice_acceptor_variant,intron_variant                                                   31\n",
       "splice_region_variant,5_prime_UTR_variant                                                30\n",
       "inframe_deletion,splice_region_variant                                                   28\n",
       "non_coding_transcript_exon_variant                                                       28\n",
       "3_prime_UTR_variant                                                                      26\n",
       "splice_acceptor_variant,coding_sequence_variant                                          20\n",
       "stop_gained,inframe_deletion                                                             13\n",
       "frameshift_variant,stop_lost                                                             10\n",
       "start_lost,5_prime_UTR_variant                                                            7\n",
       "splice_region_variant,3_prime_UTR_variant                                                 6\n",
       "stop_lost,3_prime_UTR_variant                                                             6\n",
       "frameshift_variant,start_lost                                                             5\n",
       "frameshift_variant,start_lost,start_retained_variant                                      5\n",
       "stop_gained,frameshift_variant,splice_region_variant                                      3\n",
       "start_lost,start_retained_variant,5_prime_UTR_variant                                     2\n",
       "inframe_insertion,splice_region_variant                                                   2\n",
       "stop_retained_variant                                                                     2\n",
       "splice_acceptor_variant,coding_sequence_variant,3_prime_UTR_variant,intron_variant        2\n",
       "mature_miRNA_variant                                                                      2\n",
       "stop_gained,inframe_insertion                                                             1\n",
       "stop_gained,protein_altering_variant                                                      1\n",
       "intron_variant,NMD_transcript_variant                                                     1\n",
       "splice_region_variant,non_coding_transcript_exon_variant                                  1\n",
       "intergenic_variant                                                                        1\n",
       "splice_acceptor_variant,non_coding_transcript_variant                                     1\n",
       "splice_acceptor_variant,3_prime_UTR_variant,intron_variant,NMD_transcript_variant         1\n",
       "splice_donor_variant,5_prime_UTR_variant,intron_variant                                   1\n",
       "start_lost,splice_region_variant                                                          1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 275,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 275
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:38:24.369451Z",
     "start_time": "2024-11-14T11:38:22.716569Z"
    }
   },
   "cell_type": "code",
   "source": [
    "data_for_model['Consequence'].str.split(',')\n",
    "dummy_vars = data_for_model['Consequence'].str.split(',').explode().str.get_dummies().groupby(level=0).sum()\n",
    "data_for_model = data_for_model.join(dummy_vars)\n"
   ],
   "outputs": [],
   "execution_count": 276
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:38:24.457619Z",
     "start_time": "2024-11-14T11:38:24.430391Z"
    }
   },
   "cell_type": "code",
   "source": [
    "data_for_model['Exon_Number'].value_counts()"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Exon_Number\n",
       "02/05    2012\n",
       "05/11    1835\n",
       "07/11    1565\n",
       "08/11    1510\n",
       "16/16    1468\n",
       "         ... \n",
       "28/51       1\n",
       "25/45       1\n",
       "46/87       1\n",
       "01/47       1\n",
       "03/58       1\n",
       "Name: count, Length: 1848, dtype: int64"
      ]
     },
     "execution_count": 277,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 277
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:38:24.642184Z",
     "start_time": "2024-11-14T11:38:24.598479Z"
    }
   },
   "cell_type": "code",
   "source": [
    "data_for_model.head(50)"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "     PATIENT_ID     SEX  DIAGNOSIS AGE SMOKE STATUS TMB (nonsynonymous)  \\\n",
       "0   Patient0001  Female             67      Unknown         0.333333333   \n",
       "1   Patient0001  Female             67      Unknown         0.333333333   \n",
       "2   Patient0001  Female             67      Unknown         0.333333333   \n",
       "3   Patient0001  Female             67      Unknown         0.333333333   \n",
       "4   Patient0001  Female             67      Unknown         0.333333333   \n",
       "5   Patient0001  Female             67      Unknown         0.333333333   \n",
       "6   Patient0001  Female             67      Unknown         0.333333333   \n",
       "7   Patient0001  Female             67      Unknown         0.333333333   \n",
       "8   Patient0001  Female             67      Unknown         0.333333333   \n",
       "9   Patient0001  Female             67      Unknown         0.333333333   \n",
       "10  Patient0002    Male             75      Unknown                 0.3   \n",
       "11  Patient0002    Male             75      Unknown                 0.3   \n",
       "12  Patient0002    Male             75      Unknown                 0.3   \n",
       "13  Patient0002    Male             75      Unknown                 0.3   \n",
       "14  Patient0002    Male             75      Unknown                 0.3   \n",
       "15  Patient0002    Male             75      Unknown                 0.3   \n",
       "16  Patient0002    Male             75      Unknown                 0.3   \n",
       "17  Patient0002    Male             75      Unknown                 0.3   \n",
       "18  Patient0002    Male             75      Unknown                 0.3   \n",
       "19  Patient0003  Female             45      Unknown                 0.5   \n",
       "20  Patient0003  Female             45      Unknown                 0.5   \n",
       "21  Patient0003  Female             45      Unknown                 0.5   \n",
       "22  Patient0003  Female             45      Unknown                 0.5   \n",
       "23  Patient0003  Female             45      Unknown                 0.5   \n",
       "24  Patient0003  Female             45      Unknown                 0.5   \n",
       "25  Patient0003  Female             45      Unknown                 0.5   \n",
       "26  Patient0003  Female             45      Unknown                 0.5   \n",
       "27  Patient0003  Female             45      Unknown                 0.5   \n",
       "28  Patient0003  Female             45      Unknown                 0.5   \n",
       "29  Patient0003  Female             45      Unknown                 0.5   \n",
       "30  Patient0003  Female             45      Unknown                 0.5   \n",
       "31  Patient0003  Female             45      Unknown                 0.5   \n",
       "32  Patient0003  Female             45      Unknown                 0.5   \n",
       "33  Patient0003  Female             45      Unknown                 0.5   \n",
       "34  Patient0003  Female             45      Unknown                 0.5   \n",
       "35  Patient0004    Male             70      Unknown         7.266666667   \n",
       "36  Patient0004    Male             70      Unknown         7.266666667   \n",
       "37  Patient0004    Male             70      Unknown         7.266666667   \n",
       "38  Patient0004    Male             70      Unknown         7.266666667   \n",
       "39  Patient0004    Male             70      Unknown         7.266666667   \n",
       "40  Patient0004    Male             70      Unknown         7.266666667   \n",
       "41  Patient0004    Male             70      Unknown         7.266666667   \n",
       "42  Patient0004    Male             70      Unknown         7.266666667   \n",
       "43  Patient0004    Male             70      Unknown         7.266666667   \n",
       "44  Patient0004    Male             70      Unknown         7.266666667   \n",
       "45  Patient0004    Male             70      Unknown         7.266666667   \n",
       "46  Patient0004    Male             70      Unknown         7.266666667   \n",
       "47  Patient0004    Male             70      Unknown         7.266666667   \n",
       "48  Patient0004    Male             70      Unknown         7.266666667   \n",
       "49  Patient0004    Male             70      Unknown         7.266666667   \n",
       "\n",
       "     Hugo_Symbol Chromosome  Start_Position  End_Position  \\\n",
       "0          KMT2C          7     151836340.0   151836340.0   \n",
       "1           KRAS         12      25398285.0    25398285.0   \n",
       "2           SOX9         17      70119705.0    70119705.0   \n",
       "3            APC          5     112128143.0   112128143.0   \n",
       "4            APC          5     112175147.0   112175147.0   \n",
       "5            QKI          6     163836364.0   163836364.0   \n",
       "6          FBXW7          4     153247289.0   153247289.0   \n",
       "7         PTPN11         12     112910733.0   112910736.0   \n",
       "8           GLI1         12      57863367.0    57863367.0   \n",
       "9            ATM         11     108142138.0   108142138.0   \n",
       "10          SOX9         17      70120165.0    70120165.0   \n",
       "11         EPHB1          3     134920413.0   134920413.0   \n",
       "12         FBXW7          4     153245512.0   153245512.0   \n",
       "13           APC          5     112175609.0   112175609.0   \n",
       "14          RARA         17      38508182.0    38508184.0   \n",
       "15         FBXW7          4     153249384.0   153249384.0   \n",
       "16          KRAS         12      25398284.0    25398284.0   \n",
       "17          TP53         17       7578406.0     7578406.0   \n",
       "18           APC          5     112116592.0   112116592.0   \n",
       "19           APC          5     112175639.0   112175639.0   \n",
       "20         PREX2          8      68992667.0    68992667.0   \n",
       "21           APC          5     112173926.0   112173926.0   \n",
       "22          ETV5          3     185766570.0   185766570.0   \n",
       "23      GNAS-AS1         20      57415892.0    57415892.0   \n",
       "24         AMER1          X      63411678.0    63411678.0   \n",
       "25         LRP1B          2     141031993.0   141031993.0   \n",
       "26          FAT4          4     126372813.0   126372813.0   \n",
       "27          KRAS         12      25378561.0    25378561.0   \n",
       "28          TPMT          6      18130988.0    18130988.0   \n",
       "29         ERBB2         17      37864790.0    37864790.0   \n",
       "30         FBXW7          4     153245391.0   153245391.0   \n",
       "31         SMAD3         15      67477150.0    67477151.0   \n",
       "32         CCND3          6      41903700.0    41903702.0   \n",
       "33          TP53         17       7578280.0     7578280.0   \n",
       "34  RP11-435O5.4          9      98278984.0    98278984.0   \n",
       "35         ERBB2         17      37883720.0    37883720.0   \n",
       "36        GRIN2A         16       9923440.0     9923440.0   \n",
       "37          MSH6          2      48026731.0    48026731.0   \n",
       "38         KMT2D         12      49435206.0    49435206.0   \n",
       "39        RANBP2          2     109384479.0   109384479.0   \n",
       "40        ACVR1B         12      52387796.0    52387796.0   \n",
       "41        ACVR1B         12      52369228.0    52369228.0   \n",
       "42        RANBP2          2     109379847.0   109379847.0   \n",
       "43          GLI2          2     121747649.0   121747649.0   \n",
       "44         KMT2D         12      49433721.0    49433721.0   \n",
       "45         PALB2         16      23641709.0    23641709.0   \n",
       "46         KMT2D         12      49438203.0    49438203.0   \n",
       "47         SPTA1          1     158604371.0   158604371.0   \n",
       "48        ACVR1B         12      52369073.0    52369073.0   \n",
       "49          TET3          2      74326584.0    74326584.0   \n",
       "\n",
       "                                     Consequence  ... splice_acceptor_variant  \\\n",
       "0                               missense_variant  ...                       0   \n",
       "1                               missense_variant  ...                       0   \n",
       "2                             frameshift_variant  ...                       0   \n",
       "3              stop_gained,splice_region_variant  ...                       0   \n",
       "4                                    stop_gained  ...                       0   \n",
       "5                               missense_variant  ...                       0   \n",
       "6                               missense_variant  ...                       0   \n",
       "7                                 intron_variant  ...                       0   \n",
       "8                               missense_variant  ...                       0   \n",
       "9           splice_region_variant,intron_variant  ...                       0   \n",
       "10                            frameshift_variant  ...                       0   \n",
       "11                              missense_variant  ...                       0   \n",
       "12                              missense_variant  ...                       0   \n",
       "13                            frameshift_variant  ...                       0   \n",
       "14                              inframe_deletion  ...                       0   \n",
       "15                              missense_variant  ...                       0   \n",
       "16                              missense_variant  ...                       0   \n",
       "17                              missense_variant  ...                       0   \n",
       "18                                   stop_gained  ...                       0   \n",
       "19                                   stop_gained  ...                       0   \n",
       "20                                intron_variant  ...                       0   \n",
       "21                                   stop_gained  ...                       0   \n",
       "22                              missense_variant  ...                       0   \n",
       "23  intron_variant,non_coding_transcript_variant  ...                       0   \n",
       "24                                   stop_gained  ...                       0   \n",
       "25        missense_variant,splice_region_variant  ...                       0   \n",
       "26                              missense_variant  ...                       0   \n",
       "27                              missense_variant  ...                       0   \n",
       "28                              missense_variant  ...                       0   \n",
       "29          splice_region_variant,intron_variant  ...                       0   \n",
       "30                              missense_variant  ...                       0   \n",
       "31                            frameshift_variant  ...                       0   \n",
       "32                              inframe_deletion  ...                       0   \n",
       "33                              missense_variant  ...                       0   \n",
       "34                       downstream_gene_variant  ...                       0   \n",
       "35                              missense_variant  ...                       0   \n",
       "36                              missense_variant  ...                       0   \n",
       "37                                   stop_gained  ...                       0   \n",
       "38                              missense_variant  ...                       0   \n",
       "39                              missense_variant  ...                       0   \n",
       "40                                   stop_gained  ...                       0   \n",
       "41                              missense_variant  ...                       0   \n",
       "42                              missense_variant  ...                       0   \n",
       "43                              missense_variant  ...                       0   \n",
       "44                              missense_variant  ...                       0   \n",
       "45                              missense_variant  ...                       0   \n",
       "46                              missense_variant  ...                       0   \n",
       "47                              missense_variant  ...                       0   \n",
       "48                              missense_variant  ...                       0   \n",
       "49                              missense_variant  ...                       0   \n",
       "\n",
       "   splice_donor_variant  splice_region_variant start_lost  \\\n",
       "0                     0                      0          0   \n",
       "1                     0                      0          0   \n",
       "2                     0                      0          0   \n",
       "3                     0                      1          0   \n",
       "4                     0                      0          0   \n",
       "5                     0                      0          0   \n",
       "6                     0                      0          0   \n",
       "7                     0                      0          0   \n",
       "8                     0                      0          0   \n",
       "9                     0                      1          0   \n",
       "10                    0                      0          0   \n",
       "11                    0                      0          0   \n",
       "12                    0                      0          0   \n",
       "13                    0                      0          0   \n",
       "14                    0                      0          0   \n",
       "15                    0                      0          0   \n",
       "16                    0                      0          0   \n",
       "17                    0                      0          0   \n",
       "18                    0                      0          0   \n",
       "19                    0                      0          0   \n",
       "20                    0                      0          0   \n",
       "21                    0                      0          0   \n",
       "22                    0                      0          0   \n",
       "23                    0                      0          0   \n",
       "24                    0                      0          0   \n",
       "25                    0                      1          0   \n",
       "26                    0                      0          0   \n",
       "27                    0                      0          0   \n",
       "28                    0                      0          0   \n",
       "29                    0                      1          0   \n",
       "30                    0                      0          0   \n",
       "31                    0                      0          0   \n",
       "32                    0                      0          0   \n",
       "33                    0                      0          0   \n",
       "34                    0                      0          0   \n",
       "35                    0                      0          0   \n",
       "36                    0                      0          0   \n",
       "37                    0                      0          0   \n",
       "38                    0                      0          0   \n",
       "39                    0                      0          0   \n",
       "40                    0                      0          0   \n",
       "41                    0                      0          0   \n",
       "42                    0                      0          0   \n",
       "43                    0                      0          0   \n",
       "44                    0                      0          0   \n",
       "45                    0                      0          0   \n",
       "46                    0                      0          0   \n",
       "47                    0                      0          0   \n",
       "48                    0                      0          0   \n",
       "49                    0                      0          0   \n",
       "\n",
       "   start_retained_variant stop_gained stop_lost stop_retained_variant  \\\n",
       "0                       0           0         0                     0   \n",
       "1                       0           0         0                     0   \n",
       "2                       0           0         0                     0   \n",
       "3                       0           1         0                     0   \n",
       "4                       0           1         0                     0   \n",
       "5                       0           0         0                     0   \n",
       "6                       0           0         0                     0   \n",
       "7                       0           0         0                     0   \n",
       "8                       0           0         0                     0   \n",
       "9                       0           0         0                     0   \n",
       "10                      0           0         0                     0   \n",
       "11                      0           0         0                     0   \n",
       "12                      0           0         0                     0   \n",
       "13                      0           0         0                     0   \n",
       "14                      0           0         0                     0   \n",
       "15                      0           0         0                     0   \n",
       "16                      0           0         0                     0   \n",
       "17                      0           0         0                     0   \n",
       "18                      0           1         0                     0   \n",
       "19                      0           1         0                     0   \n",
       "20                      0           0         0                     0   \n",
       "21                      0           1         0                     0   \n",
       "22                      0           0         0                     0   \n",
       "23                      0           0         0                     0   \n",
       "24                      0           1         0                     0   \n",
       "25                      0           0         0                     0   \n",
       "26                      0           0         0                     0   \n",
       "27                      0           0         0                     0   \n",
       "28                      0           0         0                     0   \n",
       "29                      0           0         0                     0   \n",
       "30                      0           0         0                     0   \n",
       "31                      0           0         0                     0   \n",
       "32                      0           0         0                     0   \n",
       "33                      0           0         0                     0   \n",
       "34                      0           0         0                     0   \n",
       "35                      0           0         0                     0   \n",
       "36                      0           0         0                     0   \n",
       "37                      0           1         0                     0   \n",
       "38                      0           0         0                     0   \n",
       "39                      0           0         0                     0   \n",
       "40                      0           1         0                     0   \n",
       "41                      0           0         0                     0   \n",
       "42                      0           0         0                     0   \n",
       "43                      0           0         0                     0   \n",
       "44                      0           0         0                     0   \n",
       "45                      0           0         0                     0   \n",
       "46                      0           0         0                     0   \n",
       "47                      0           0         0                     0   \n",
       "48                      0           0         0                     0   \n",
       "49                      0           0         0                     0   \n",
       "\n",
       "   synonymous_variant  upstream_gene_variant  \n",
       "0                   0                      0  \n",
       "1                   0                      0  \n",
       "2                   0                      0  \n",
       "3                   0                      0  \n",
       "4                   0                      0  \n",
       "5                   0                      0  \n",
       "6                   0                      0  \n",
       "7                   0                      0  \n",
       "8                   0                      0  \n",
       "9                   0                      0  \n",
       "10                  0                      0  \n",
       "11                  0                      0  \n",
       "12                  0                      0  \n",
       "13                  0                      0  \n",
       "14                  0                      0  \n",
       "15                  0                      0  \n",
       "16                  0                      0  \n",
       "17                  0                      0  \n",
       "18                  0                      0  \n",
       "19                  0                      0  \n",
       "20                  0                      0  \n",
       "21                  0                      0  \n",
       "22                  0                      0  \n",
       "23                  0                      0  \n",
       "24                  0                      0  \n",
       "25                  0                      0  \n",
       "26                  0                      0  \n",
       "27                  0                      0  \n",
       "28                  0                      0  \n",
       "29                  0                      0  \n",
       "30                  0                      0  \n",
       "31                  0                      0  \n",
       "32                  0                      0  \n",
       "33                  0                      0  \n",
       "34                  0                      0  \n",
       "35                  0                      0  \n",
       "36                  0                      0  \n",
       "37                  0                      0  \n",
       "38                  0                      0  \n",
       "39                  0                      0  \n",
       "40                  0                      0  \n",
       "41                  0                      0  \n",
       "42                  0                      0  \n",
       "43                  0                      0  \n",
       "44                  0                      0  \n",
       "45                  0                      0  \n",
       "46                  0                      0  \n",
       "47                  0                      0  \n",
       "48                  0                      0  \n",
       "49                  0                      0  \n",
       "\n",
       "[50 rows x 44 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>SEX</th>\n",
       "      <th>DIAGNOSIS AGE</th>\n",
       "      <th>SMOKE STATUS</th>\n",
       "      <th>TMB (nonsynonymous)</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>...</th>\n",
       "      <th>splice_acceptor_variant</th>\n",
       "      <th>splice_donor_variant</th>\n",
       "      <th>splice_region_variant</th>\n",
       "      <th>start_lost</th>\n",
       "      <th>start_retained_variant</th>\n",
       "      <th>stop_gained</th>\n",
       "      <th>stop_lost</th>\n",
       "      <th>stop_retained_variant</th>\n",
       "      <th>synonymous_variant</th>\n",
       "      <th>upstream_gene_variant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>7</td>\n",
       "      <td>151836340.0</td>\n",
       "      <td>151836340.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>12</td>\n",
       "      <td>25398285.0</td>\n",
       "      <td>25398285.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>SOX9</td>\n",
       "      <td>17</td>\n",
       "      <td>70119705.0</td>\n",
       "      <td>70119705.0</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112128143.0</td>\n",
       "      <td>112128143.0</td>\n",
       "      <td>stop_gained,splice_region_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175147.0</td>\n",
       "      <td>112175147.0</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>QKI</td>\n",
       "      <td>6</td>\n",
       "      <td>163836364.0</td>\n",
       "      <td>163836364.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>FBXW7</td>\n",
       "      <td>4</td>\n",
       "      <td>153247289.0</td>\n",
       "      <td>153247289.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>PTPN11</td>\n",
       "      <td>12</td>\n",
       "      <td>112910733.0</td>\n",
       "      <td>112910736.0</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>GLI1</td>\n",
       "      <td>12</td>\n",
       "      <td>57863367.0</td>\n",
       "      <td>57863367.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>ATM</td>\n",
       "      <td>11</td>\n",
       "      <td>108142138.0</td>\n",
       "      <td>108142138.0</td>\n",
       "      <td>splice_region_variant,intron_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>SOX9</td>\n",
       "      <td>17</td>\n",
       "      <td>70120165.0</td>\n",
       "      <td>70120165.0</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>EPHB1</td>\n",
       "      <td>3</td>\n",
       "      <td>134920413.0</td>\n",
       "      <td>134920413.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>FBXW7</td>\n",
       "      <td>4</td>\n",
       "      <td>153245512.0</td>\n",
       "      <td>153245512.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175609.0</td>\n",
       "      <td>112175609.0</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>RARA</td>\n",
       "      <td>17</td>\n",
       "      <td>38508182.0</td>\n",
       "      <td>38508184.0</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>FBXW7</td>\n",
       "      <td>4</td>\n",
       "      <td>153249384.0</td>\n",
       "      <td>153249384.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>12</td>\n",
       "      <td>25398284.0</td>\n",
       "      <td>25398284.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578406.0</td>\n",
       "      <td>7578406.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112116592.0</td>\n",
       "      <td>112116592.0</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Patient0003</td>\n",
       "      <td>Female</td>\n",
       "      <td>45</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.5</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112175639.0</td>\n",
       "      <td>112175639.0</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Patient0003</td>\n",
       "      <td>Female</td>\n",
       "      <td>45</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.5</td>\n",
       "      <td>PREX2</td>\n",
       "      <td>8</td>\n",
       "      <td>68992667.0</td>\n",
       "      <td>68992667.0</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Patient0003</td>\n",
       "      <td>Female</td>\n",
       "      <td>45</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.5</td>\n",
       "      <td>APC</td>\n",
       "      <td>5</td>\n",
       "      <td>112173926.0</td>\n",
       "      <td>112173926.0</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Patient0003</td>\n",
       "      <td>Female</td>\n",
       "      <td>45</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.5</td>\n",
       "      <td>ETV5</td>\n",
       "      <td>3</td>\n",
       "      <td>185766570.0</td>\n",
       "      <td>185766570.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Patient0003</td>\n",
       "      <td>Female</td>\n",
       "      <td>45</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.5</td>\n",
       "      <td>GNAS-AS1</td>\n",
       "      <td>20</td>\n",
       "      <td>57415892.0</td>\n",
       "      <td>57415892.0</td>\n",
       "      <td>intron_variant,non_coding_transcript_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Patient0003</td>\n",
       "      <td>Female</td>\n",
       "      <td>45</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.5</td>\n",
       "      <td>AMER1</td>\n",
       "      <td>X</td>\n",
       "      <td>63411678.0</td>\n",
       "      <td>63411678.0</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Patient0003</td>\n",
       "      <td>Female</td>\n",
       "      <td>45</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.5</td>\n",
       "      <td>LRP1B</td>\n",
       "      <td>2</td>\n",
       "      <td>141031993.0</td>\n",
       "      <td>141031993.0</td>\n",
       "      <td>missense_variant,splice_region_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Patient0003</td>\n",
       "      <td>Female</td>\n",
       "      <td>45</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.5</td>\n",
       "      <td>FAT4</td>\n",
       "      <td>4</td>\n",
       "      <td>126372813.0</td>\n",
       "      <td>126372813.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Patient0003</td>\n",
       "      <td>Female</td>\n",
       "      <td>45</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.5</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>12</td>\n",
       "      <td>25378561.0</td>\n",
       "      <td>25378561.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Patient0003</td>\n",
       "      <td>Female</td>\n",
       "      <td>45</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.5</td>\n",
       "      <td>TPMT</td>\n",
       "      <td>6</td>\n",
       "      <td>18130988.0</td>\n",
       "      <td>18130988.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Patient0003</td>\n",
       "      <td>Female</td>\n",
       "      <td>45</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.5</td>\n",
       "      <td>ERBB2</td>\n",
       "      <td>17</td>\n",
       "      <td>37864790.0</td>\n",
       "      <td>37864790.0</td>\n",
       "      <td>splice_region_variant,intron_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Patient0003</td>\n",
       "      <td>Female</td>\n",
       "      <td>45</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.5</td>\n",
       "      <td>FBXW7</td>\n",
       "      <td>4</td>\n",
       "      <td>153245391.0</td>\n",
       "      <td>153245391.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Patient0003</td>\n",
       "      <td>Female</td>\n",
       "      <td>45</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.5</td>\n",
       "      <td>SMAD3</td>\n",
       "      <td>15</td>\n",
       "      <td>67477150.0</td>\n",
       "      <td>67477151.0</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Patient0003</td>\n",
       "      <td>Female</td>\n",
       "      <td>45</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.5</td>\n",
       "      <td>CCND3</td>\n",
       "      <td>6</td>\n",
       "      <td>41903700.0</td>\n",
       "      <td>41903702.0</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Patient0003</td>\n",
       "      <td>Female</td>\n",
       "      <td>45</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.5</td>\n",
       "      <td>TP53</td>\n",
       "      <td>17</td>\n",
       "      <td>7578280.0</td>\n",
       "      <td>7578280.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Patient0003</td>\n",
       "      <td>Female</td>\n",
       "      <td>45</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.5</td>\n",
       "      <td>RP11-435O5.4</td>\n",
       "      <td>9</td>\n",
       "      <td>98278984.0</td>\n",
       "      <td>98278984.0</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Patient0004</td>\n",
       "      <td>Male</td>\n",
       "      <td>70</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7.266666667</td>\n",
       "      <td>ERBB2</td>\n",
       "      <td>17</td>\n",
       "      <td>37883720.0</td>\n",
       "      <td>37883720.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Patient0004</td>\n",
       "      <td>Male</td>\n",
       "      <td>70</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7.266666667</td>\n",
       "      <td>GRIN2A</td>\n",
       "      <td>16</td>\n",
       "      <td>9923440.0</td>\n",
       "      <td>9923440.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Patient0004</td>\n",
       "      <td>Male</td>\n",
       "      <td>70</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7.266666667</td>\n",
       "      <td>MSH6</td>\n",
       "      <td>2</td>\n",
       "      <td>48026731.0</td>\n",
       "      <td>48026731.0</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Patient0004</td>\n",
       "      <td>Male</td>\n",
       "      <td>70</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7.266666667</td>\n",
       "      <td>KMT2D</td>\n",
       "      <td>12</td>\n",
       "      <td>49435206.0</td>\n",
       "      <td>49435206.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Patient0004</td>\n",
       "      <td>Male</td>\n",
       "      <td>70</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7.266666667</td>\n",
       "      <td>RANBP2</td>\n",
       "      <td>2</td>\n",
       "      <td>109384479.0</td>\n",
       "      <td>109384479.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Patient0004</td>\n",
       "      <td>Male</td>\n",
       "      <td>70</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7.266666667</td>\n",
       "      <td>ACVR1B</td>\n",
       "      <td>12</td>\n",
       "      <td>52387796.0</td>\n",
       "      <td>52387796.0</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Patient0004</td>\n",
       "      <td>Male</td>\n",
       "      <td>70</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7.266666667</td>\n",
       "      <td>ACVR1B</td>\n",
       "      <td>12</td>\n",
       "      <td>52369228.0</td>\n",
       "      <td>52369228.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Patient0004</td>\n",
       "      <td>Male</td>\n",
       "      <td>70</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7.266666667</td>\n",
       "      <td>RANBP2</td>\n",
       "      <td>2</td>\n",
       "      <td>109379847.0</td>\n",
       "      <td>109379847.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Patient0004</td>\n",
       "      <td>Male</td>\n",
       "      <td>70</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7.266666667</td>\n",
       "      <td>GLI2</td>\n",
       "      <td>2</td>\n",
       "      <td>121747649.0</td>\n",
       "      <td>121747649.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Patient0004</td>\n",
       "      <td>Male</td>\n",
       "      <td>70</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7.266666667</td>\n",
       "      <td>KMT2D</td>\n",
       "      <td>12</td>\n",
       "      <td>49433721.0</td>\n",
       "      <td>49433721.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Patient0004</td>\n",
       "      <td>Male</td>\n",
       "      <td>70</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7.266666667</td>\n",
       "      <td>PALB2</td>\n",
       "      <td>16</td>\n",
       "      <td>23641709.0</td>\n",
       "      <td>23641709.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Patient0004</td>\n",
       "      <td>Male</td>\n",
       "      <td>70</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7.266666667</td>\n",
       "      <td>KMT2D</td>\n",
       "      <td>12</td>\n",
       "      <td>49438203.0</td>\n",
       "      <td>49438203.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Patient0004</td>\n",
       "      <td>Male</td>\n",
       "      <td>70</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7.266666667</td>\n",
       "      <td>SPTA1</td>\n",
       "      <td>1</td>\n",
       "      <td>158604371.0</td>\n",
       "      <td>158604371.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Patient0004</td>\n",
       "      <td>Male</td>\n",
       "      <td>70</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7.266666667</td>\n",
       "      <td>ACVR1B</td>\n",
       "      <td>12</td>\n",
       "      <td>52369073.0</td>\n",
       "      <td>52369073.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Patient0004</td>\n",
       "      <td>Male</td>\n",
       "      <td>70</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>7.266666667</td>\n",
       "      <td>TET3</td>\n",
       "      <td>2</td>\n",
       "      <td>74326584.0</td>\n",
       "      <td>74326584.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>50 rows  44 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 278,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 278
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:38:24.930541Z",
     "start_time": "2024-11-14T11:38:24.750316Z"
    }
   },
   "cell_type": "code",
   "source": [
    "data_for_model = pd.get_dummies(data_for_model, columns=['SMOKE STATUS', 'SEX', 'VAR_TYPE_SX', 'Chromosome', 'Variant_Type'], drop_first=False)\n",
    "data_for_model.drop('SMOKE STATUS_Unknown', axis=1, inplace=True)"
   ],
   "outputs": [],
   "execution_count": 279
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:38:36.002068Z",
     "start_time": "2024-11-14T11:38:35.929990Z"
    }
   },
   "cell_type": "code",
   "source": [
    "#data_for_model['HGVSc'].str.replace(r'\\.\\d', '', regex=True).value_counts()\n"
   ],
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'HGVSc'",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mKeyError\u001B[0m                                  Traceback (most recent call last)",
      "File \u001B[0;32m~/Sites/llm-project/venv/lib/python3.10/site-packages/pandas/core/indexes/base.py:3805\u001B[0m, in \u001B[0;36mIndex.get_loc\u001B[0;34m(self, key)\u001B[0m\n\u001B[1;32m   3804\u001B[0m \u001B[38;5;28;01mtry\u001B[39;00m:\n\u001B[0;32m-> 3805\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43m_engine\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mget_loc\u001B[49m\u001B[43m(\u001B[49m\u001B[43mcasted_key\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m   3806\u001B[0m \u001B[38;5;28;01mexcept\u001B[39;00m \u001B[38;5;167;01mKeyError\u001B[39;00m \u001B[38;5;28;01mas\u001B[39;00m err:\n",
      "File \u001B[0;32mindex.pyx:167\u001B[0m, in \u001B[0;36mpandas._libs.index.IndexEngine.get_loc\u001B[0;34m()\u001B[0m\n",
      "File \u001B[0;32mindex.pyx:196\u001B[0m, in \u001B[0;36mpandas._libs.index.IndexEngine.get_loc\u001B[0;34m()\u001B[0m\n",
      "File \u001B[0;32mpandas/_libs/hashtable_class_helper.pxi:7081\u001B[0m, in \u001B[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001B[0;34m()\u001B[0m\n",
      "File \u001B[0;32mpandas/_libs/hashtable_class_helper.pxi:7089\u001B[0m, in \u001B[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001B[0;34m()\u001B[0m\n",
      "\u001B[0;31mKeyError\u001B[0m: 'HGVSc'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001B[0;31mKeyError\u001B[0m                                  Traceback (most recent call last)",
      "Cell \u001B[0;32mIn[281], line 1\u001B[0m\n\u001B[0;32m----> 1\u001B[0m \u001B[43mdata_for_model\u001B[49m\u001B[43m[\u001B[49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[38;5;124;43mHGVSc\u001B[39;49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[43m]\u001B[49m\u001B[38;5;241m.\u001B[39mstr\u001B[38;5;241m.\u001B[39mreplace(\u001B[38;5;124mr\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;124m\\\u001B[39m\u001B[38;5;124m.\u001B[39m\u001B[38;5;124m\\\u001B[39m\u001B[38;5;124md\u001B[39m\u001B[38;5;124m'\u001B[39m, \u001B[38;5;124m'\u001B[39m\u001B[38;5;124m'\u001B[39m, regex\u001B[38;5;241m=\u001B[39m\u001B[38;5;28;01mTrue\u001B[39;00m)\u001B[38;5;241m.\u001B[39mvalue_counts()\n",
      "File \u001B[0;32m~/Sites/llm-project/venv/lib/python3.10/site-packages/pandas/core/frame.py:4102\u001B[0m, in \u001B[0;36mDataFrame.__getitem__\u001B[0;34m(self, key)\u001B[0m\n\u001B[1;32m   4100\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mcolumns\u001B[38;5;241m.\u001B[39mnlevels \u001B[38;5;241m>\u001B[39m \u001B[38;5;241m1\u001B[39m:\n\u001B[1;32m   4101\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_getitem_multilevel(key)\n\u001B[0;32m-> 4102\u001B[0m indexer \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mcolumns\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mget_loc\u001B[49m\u001B[43m(\u001B[49m\u001B[43mkey\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m   4103\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m is_integer(indexer):\n\u001B[1;32m   4104\u001B[0m     indexer \u001B[38;5;241m=\u001B[39m [indexer]\n",
      "File \u001B[0;32m~/Sites/llm-project/venv/lib/python3.10/site-packages/pandas/core/indexes/base.py:3812\u001B[0m, in \u001B[0;36mIndex.get_loc\u001B[0;34m(self, key)\u001B[0m\n\u001B[1;32m   3807\u001B[0m     \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;28misinstance\u001B[39m(casted_key, \u001B[38;5;28mslice\u001B[39m) \u001B[38;5;129;01mor\u001B[39;00m (\n\u001B[1;32m   3808\u001B[0m         \u001B[38;5;28misinstance\u001B[39m(casted_key, abc\u001B[38;5;241m.\u001B[39mIterable)\n\u001B[1;32m   3809\u001B[0m         \u001B[38;5;129;01mand\u001B[39;00m \u001B[38;5;28many\u001B[39m(\u001B[38;5;28misinstance\u001B[39m(x, \u001B[38;5;28mslice\u001B[39m) \u001B[38;5;28;01mfor\u001B[39;00m x \u001B[38;5;129;01min\u001B[39;00m casted_key)\n\u001B[1;32m   3810\u001B[0m     ):\n\u001B[1;32m   3811\u001B[0m         \u001B[38;5;28;01mraise\u001B[39;00m InvalidIndexError(key)\n\u001B[0;32m-> 3812\u001B[0m     \u001B[38;5;28;01mraise\u001B[39;00m \u001B[38;5;167;01mKeyError\u001B[39;00m(key) \u001B[38;5;28;01mfrom\u001B[39;00m \u001B[38;5;21;01merr\u001B[39;00m\n\u001B[1;32m   3813\u001B[0m \u001B[38;5;28;01mexcept\u001B[39;00m \u001B[38;5;167;01mTypeError\u001B[39;00m:\n\u001B[1;32m   3814\u001B[0m     \u001B[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001B[39;00m\n\u001B[1;32m   3815\u001B[0m     \u001B[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001B[39;00m\n\u001B[1;32m   3816\u001B[0m     \u001B[38;5;66;03m#  the TypeError.\u001B[39;00m\n\u001B[1;32m   3817\u001B[0m     \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_check_indexing_error(key)\n",
      "\u001B[0;31mKeyError\u001B[0m: 'HGVSc'"
     ]
    }
   ],
   "execution_count": 281
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-14T11:38:48.008941Z",
     "start_time": "2024-11-14T11:38:47.985516Z"
    }
   },
   "cell_type": "code",
   "source": [
    "data_for_model.head()"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "    PATIENT_ID  DIAGNOSIS AGE TMB (nonsynonymous) Hugo_Symbol  Start_Position  \\\n",
       "0  Patient0001             67         0.333333333       KMT2C     151836340.0   \n",
       "1  Patient0001             67         0.333333333        KRAS      25398285.0   \n",
       "2  Patient0001             67         0.333333333        SOX9      70119705.0   \n",
       "3  Patient0001             67         0.333333333         APC     112128143.0   \n",
       "4  Patient0001             67         0.333333333         APC     112175147.0   \n",
       "\n",
       "   End_Position                        Consequence SNP_event  \\\n",
       "0   151836340.0                   missense_variant       C>T   \n",
       "1    25398285.0                   missense_variant       C>A   \n",
       "2    70119705.0                 frameshift_variant       C>-   \n",
       "3   112128143.0  stop_gained,splice_region_variant       C>T   \n",
       "4   112175147.0                        stop_gained       G>T   \n",
       "\n",
       "   Protein_position   Codons  ... Chromosome_7 Chromosome_8 Chromosome_9  \\\n",
       "0            4822.0  cGt/cAt  ...         True        False        False   \n",
       "1              12.0  Ggt/Tgt  ...        False        False        False   \n",
       "2             236.0   aCc/ac  ...        False        False        False   \n",
       "3             216.0  Cga/Tga  ...        False        False        False   \n",
       "4            1286.0  Gaa/Taa  ...        False        False        False   \n",
       "\n",
       "  Chromosome_X  Variant_Type_DEL  Variant_Type_DNP  Variant_Type_INS  \\\n",
       "0        False             False             False             False   \n",
       "1        False             False             False             False   \n",
       "2        False              True             False             False   \n",
       "3        False             False             False             False   \n",
       "4        False             False             False             False   \n",
       "\n",
       "   Variant_Type_ONP  Variant_Type_SNP  Variant_Type_TNP  \n",
       "0             False              True             False  \n",
       "1             False              True             False  \n",
       "2             False             False             False  \n",
       "3             False              True             False  \n",
       "4             False              True             False  \n",
       "\n",
       "[5 rows x 74 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>DIAGNOSIS AGE</th>\n",
       "      <th>TMB (nonsynonymous)</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>SNP_event</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Codons</th>\n",
       "      <th>...</th>\n",
       "      <th>Chromosome_7</th>\n",
       "      <th>Chromosome_8</th>\n",
       "      <th>Chromosome_9</th>\n",
       "      <th>Chromosome_X</th>\n",
       "      <th>Variant_Type_DEL</th>\n",
       "      <th>Variant_Type_DNP</th>\n",
       "      <th>Variant_Type_INS</th>\n",
       "      <th>Variant_Type_ONP</th>\n",
       "      <th>Variant_Type_SNP</th>\n",
       "      <th>Variant_Type_TNP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>67</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>151836340.0</td>\n",
       "      <td>151836340.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>4822.0</td>\n",
       "      <td>cGt/cAt</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>67</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>25398285.0</td>\n",
       "      <td>25398285.0</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>12.0</td>\n",
       "      <td>Ggt/Tgt</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>67</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>SOX9</td>\n",
       "      <td>70119705.0</td>\n",
       "      <td>70119705.0</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>C&gt;-</td>\n",
       "      <td>236.0</td>\n",
       "      <td>aCc/ac</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>67</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>APC</td>\n",
       "      <td>112128143.0</td>\n",
       "      <td>112128143.0</td>\n",
       "      <td>stop_gained,splice_region_variant</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>216.0</td>\n",
       "      <td>Cga/Tga</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>67</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>APC</td>\n",
       "      <td>112175147.0</td>\n",
       "      <td>112175147.0</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>G&gt;T</td>\n",
       "      <td>1286.0</td>\n",
       "      <td>Gaa/Taa</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  74 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 282,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 282
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
