{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "#  DataBase contruction"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-05T13:52:50.258888Z",
     "start_time": "2025-05-05T13:52:50.252578Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import pandas as pd\n",
    "import itertools\n",
    "from itertools import combinations\n",
    "import os\n",
    "import glob\n",
    "from pandas.core.nanops import nanall\n",
    "from sklearn.preprocessing import LabelEncoder, OneHotEncoder\n",
    "from sklearn import tree\n",
    "from XGBoost_Model import *\n",
    "import sns\n",
    "from matplotlib import pyplot as plt\n",
    "from scipy.stats import chi2_contingency"
   ],
   "outputs": [],
   "execution_count": 2
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-05T13:52:51.890519Z",
     "start_time": "2025-05-05T13:52:51.313023Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# reading all the data files\n",
    "data_clinical_patient = pd.read_csv('pan_origimed_2020/data_clinical_patient.txt', sep=\"\\t\")\n",
    "data_clinical_sample = pd.read_csv('pan_origimed_2020/data_clinical_sample.txt', sep=\"\\t\")\n",
    "data_cna_log2 = pd.read_csv('pan_origimed_2020/data_cna_log2.txt', sep=\"\\t\")\n",
    "data_cna = pd.read_csv('pan_origimed_2020/data_cna.txt', sep=\"\\t\")\n",
    "data_mutations = pd.read_csv('pan_origimed_2020/data_mutations.txt', sep=\"\\t\", header=2, dtype={\"Exon_Number\": \"string\"})\n",
    "data_sv = pd.read_csv('pan_origimed_2020/data_sv.txt', sep=\"\\t\")"
   ],
   "outputs": [],
   "execution_count": 3
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-05T13:52:51.948583Z",
     "start_time": "2025-05-05T13:52:51.946963Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# removing bad rows\n",
    "data_clinical_sample = data_clinical_sample[4:]\n",
    "data_clinical_patient = data_clinical_patient[4:]"
   ],
   "outputs": [],
   "execution_count": 4
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-05T13:52:52.539256Z",
     "start_time": "2025-05-05T13:52:52.536884Z"
    }
   },
   "cell_type": "code",
   "source": "data_clinical_patient = data_clinical_patient.rename(columns={'#Patient Identifier': 'PATIENT_ID'})",
   "outputs": [],
   "execution_count": 5
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-05T13:52:53.233694Z",
     "start_time": "2025-05-05T13:52:53.227326Z"
    }
   },
   "cell_type": "code",
   "source": "data_clinical_patient.head()",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "    PATIENT_ID     Sex Diagnosis Age Smoke Status         Treatment\n",
       "4  Patient0001  Female            67      Unknown  Other_Treatments\n",
       "5  Patient0002    Male            75      Unknown   Treatment-naive\n",
       "6  Patient0003  Female            45      Unknown   Treatment-naive\n",
       "7  Patient0004    Male            70      Unknown   Treatment-naive\n",
       "8  Patient0005    Male            53      Unknown   Treatment-naive"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Diagnosis Age</th>\n",
       "      <th>Smoke Status</th>\n",
       "      <th>Treatment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Other_Treatments</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Treatment-naive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Patient0003</td>\n",
       "      <td>Female</td>\n",
       "      <td>45</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Treatment-naive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Patient0004</td>\n",
       "      <td>Male</td>\n",
       "      <td>70</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Treatment-naive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Patient0005</td>\n",
       "      <td>Male</td>\n",
       "      <td>53</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Treatment-naive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 6
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-05T13:52:54.135565Z",
     "start_time": "2025-05-05T13:52:54.131149Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# matching the sample id to match other tables\n",
    "data_clinical_patient[\"SAMPLE_ID\"] = data_clinical_patient[\"PATIENT_ID\"].apply(lambda x: \"P-\" + x[7:])"
   ],
   "outputs": [],
   "execution_count": 7
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-05T13:52:54.891402Z",
     "start_time": "2025-05-05T13:52:54.887927Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# make all sample id header name the same - \"SAMPLE_ID\"\n",
    "data_clinical_sample.rename(columns={\"Sample Identifier\": 'SAMPLE_ID'}, inplace=True)\n",
    "data_mutations.rename(columns={\"Tumor_Sample_Barcode\": 'SAMPLE_ID'}, inplace=True)\n",
    "data_sv.rename(columns={\"Sample_Id\": 'SAMPLE_ID'}, inplace=True)"
   ],
   "outputs": [],
   "execution_count": 8
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-05T13:52:55.743233Z",
     "start_time": "2025-05-05T13:52:55.566143Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# merge everything\n",
    "merged_clinical_data = data_clinical_patient.merge(data_clinical_sample, on=\"SAMPLE_ID\", how='outer')\n",
    "merged_mutations_data = merged_clinical_data.merge(data_mutations, on=\"SAMPLE_ID\", how='outer')\n",
    "merged_all_data = merged_mutations_data.merge(data_sv, on=\"SAMPLE_ID\", how='outer')"
   ],
   "outputs": [],
   "execution_count": 9
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-05T13:53:01.813231Z",
     "start_time": "2025-05-05T13:53:01.797283Z"
    }
   },
   "cell_type": "code",
   "source": "merged_all_data[\"SNP_event\"] = merged_all_data[\"Reference_Allele\"].fillna(\"\").astype(str) + \">\" + merged_all_data[\"Tumor_Seq_Allele2\"].fillna(\"\").astype(str)\n",
   "outputs": [],
   "execution_count": 10
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-05T13:53:02.595823Z",
     "start_time": "2025-05-05T13:53:02.579064Z"
    }
   },
   "cell_type": "code",
   "source": [
    "data_for_model = merged_all_data[[\"PATIENT_ID\", \"Cancer Type\", 'Cancer Type Detailed', 'Tumor Stage',\n",
    "                                'Sample Type', \"Sex\", \"Diagnosis Age\", \"Smoke Status\", \"TMB (nonsynonymous)\",\n",
    "                                \"Hugo_Symbol\", \"Chromosome\", \"Start_Position\", \"End_Position\",\n",
    "                                \"Consequence\", \"Variant_Type\", \"SNP_event\", \"Protein_position\", \"Codons\",\n",
    "                                \"Exon_Number\",\"VAR_TYPE_SX\", \"Site1_Hugo_Symbol\", \"Site2_Hugo_Symbol\",\"Event_Info\"]]"
   ],
   "outputs": [],
   "execution_count": 11
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-05T13:53:03.653734Z",
     "start_time": "2025-05-05T13:53:03.649512Z"
    }
   },
   "cell_type": "code",
   "source": "data_for_model[\"Exon_Number\"].isnull().sum()",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(15278)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 12
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-05T13:53:04.436626Z",
     "start_time": "2025-05-05T13:53:04.427434Z"
    }
   },
   "cell_type": "code",
   "source": "data_for_model.head(20)",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "     PATIENT_ID           Cancer Type       Cancer Type Detailed Tumor Stage  \\\n",
       "0   Patient0001  Colorectal Carcinoma  Colorectal Adenocarcinoma          IV   \n",
       "1   Patient0001  Colorectal Carcinoma  Colorectal Adenocarcinoma          IV   \n",
       "2   Patient0001  Colorectal Carcinoma  Colorectal Adenocarcinoma          IV   \n",
       "3   Patient0001  Colorectal Carcinoma  Colorectal Adenocarcinoma          IV   \n",
       "4   Patient0001  Colorectal Carcinoma  Colorectal Adenocarcinoma          IV   \n",
       "5   Patient0001  Colorectal Carcinoma  Colorectal Adenocarcinoma          IV   \n",
       "6   Patient0001  Colorectal Carcinoma  Colorectal Adenocarcinoma          IV   \n",
       "7   Patient0001  Colorectal Carcinoma  Colorectal Adenocarcinoma          IV   \n",
       "8   Patient0001  Colorectal Carcinoma  Colorectal Adenocarcinoma          IV   \n",
       "9   Patient0001  Colorectal Carcinoma  Colorectal Adenocarcinoma          IV   \n",
       "10  Patient0002  Colorectal Carcinoma  Colorectal Adenocarcinoma          II   \n",
       "11  Patient0002  Colorectal Carcinoma  Colorectal Adenocarcinoma          II   \n",
       "12  Patient0002  Colorectal Carcinoma  Colorectal Adenocarcinoma          II   \n",
       "13  Patient0002  Colorectal Carcinoma  Colorectal Adenocarcinoma          II   \n",
       "14  Patient0002  Colorectal Carcinoma  Colorectal Adenocarcinoma          II   \n",
       "15  Patient0002  Colorectal Carcinoma  Colorectal Adenocarcinoma          II   \n",
       "16  Patient0002  Colorectal Carcinoma  Colorectal Adenocarcinoma          II   \n",
       "17  Patient0002  Colorectal Carcinoma  Colorectal Adenocarcinoma          II   \n",
       "18  Patient0002  Colorectal Carcinoma  Colorectal Adenocarcinoma          II   \n",
       "19  Patient0003  Colorectal Carcinoma  Colorectal Adenocarcinoma         III   \n",
       "\n",
       "   Sample Type     Sex Diagnosis Age Smoke Status TMB (nonsynonymous)  \\\n",
       "0      Primary  Female            67      Unknown         0.333333333   \n",
       "1      Primary  Female            67      Unknown         0.333333333   \n",
       "2      Primary  Female            67      Unknown         0.333333333   \n",
       "3      Primary  Female            67      Unknown         0.333333333   \n",
       "4      Primary  Female            67      Unknown         0.333333333   \n",
       "5      Primary  Female            67      Unknown         0.333333333   \n",
       "6      Primary  Female            67      Unknown         0.333333333   \n",
       "7      Primary  Female            67      Unknown         0.333333333   \n",
       "8      Primary  Female            67      Unknown         0.333333333   \n",
       "9      Primary  Female            67      Unknown         0.333333333   \n",
       "10     Primary    Male            75      Unknown                 0.3   \n",
       "11     Primary    Male            75      Unknown                 0.3   \n",
       "12     Primary    Male            75      Unknown                 0.3   \n",
       "13     Primary    Male            75      Unknown                 0.3   \n",
       "14     Primary    Male            75      Unknown                 0.3   \n",
       "15     Primary    Male            75      Unknown                 0.3   \n",
       "16     Primary    Male            75      Unknown                 0.3   \n",
       "17     Primary    Male            75      Unknown                 0.3   \n",
       "18     Primary    Male            75      Unknown                 0.3   \n",
       "19     Primary  Female            45      Unknown                 0.5   \n",
       "\n",
       "   Hugo_Symbol  ...                           Consequence  Variant_Type  \\\n",
       "0        KMT2C  ...                      missense_variant           SNP   \n",
       "1         KRAS  ...                      missense_variant           SNP   \n",
       "2         SOX9  ...                    frameshift_variant           DEL   \n",
       "3          APC  ...     stop_gained,splice_region_variant           SNP   \n",
       "4          APC  ...                           stop_gained           SNP   \n",
       "5          QKI  ...                      missense_variant           SNP   \n",
       "6        FBXW7  ...                      missense_variant           SNP   \n",
       "7       PTPN11  ...                        intron_variant           DEL   \n",
       "8         GLI1  ...                      missense_variant           SNP   \n",
       "9          ATM  ...  splice_region_variant,intron_variant           SNP   \n",
       "10        SOX9  ...                    frameshift_variant           DEL   \n",
       "11       EPHB1  ...                      missense_variant           SNP   \n",
       "12       FBXW7  ...                      missense_variant           SNP   \n",
       "13         APC  ...                    frameshift_variant           DEL   \n",
       "14        RARA  ...                      inframe_deletion           DEL   \n",
       "15       FBXW7  ...                      missense_variant           SNP   \n",
       "16        KRAS  ...                      missense_variant           SNP   \n",
       "17        TP53  ...                      missense_variant           SNP   \n",
       "18         APC  ...                           stop_gained           SNP   \n",
       "19         APC  ...                           stop_gained           SNP   \n",
       "\n",
       "    SNP_event Protein_position   Codons Exon_Number         VAR_TYPE_SX  \\\n",
       "0         C>T           4822.0  cGt/cAt       57/59  Substitution/Indel   \n",
       "1         C>A             12.0  Ggt/Tgt       5-Feb  Substitution/Indel   \n",
       "2         C>-            236.0   aCc/ac       3-Mar          Truncation   \n",
       "3         C>T            216.0  Cga/Tga      16-Jul          Truncation   \n",
       "4         G>T           1286.0  Gaa/Taa       16/16          Truncation   \n",
       "5         G>A             47.0  Gaa/Aaa       8-Jan  Substitution/Indel   \n",
       "6         G>A            505.0  Cgc/Tgc      12-Oct  Substitution/Indel   \n",
       "7      TTTC>-              NaN      NaN        <NA>  Substitution/Indel   \n",
       "8         G>T            488.0  Gct/Tct      12-Nov  Substitution/Indel   \n",
       "9         G>A           1026.0      NaN        <NA>  Substitution/Indel   \n",
       "10        G>-            389.0   ccG/cc       3-Mar          Truncation   \n",
       "11        G>A            743.0  cGg/cAg      16-Dec  Substitution/Indel   \n",
       "12        T>C            560.0  gAt/gGt      12-Nov  Substitution/Indel   \n",
       "13        C>-           1440.0   Cca/ca       16/16          Truncation   \n",
       "14      AAG>-            164.0    AAG/-       9-May  Substitution/Indel   \n",
       "15        C>T            465.0  cGt/cAt      12-Sep  Substitution/Indel   \n",
       "16        C>A             12.0  gGt/gTt       5-Feb  Substitution/Indel   \n",
       "17        C>T            175.0  cGc/cAc      11-May  Substitution/Indel   \n",
       "18        C>T            213.0  Cga/Tga      16-Jun          Truncation   \n",
       "19        C>T           1450.0  Cga/Tga       16/16          Truncation   \n",
       "\n",
       "   Site1_Hugo_Symbol Site2_Hugo_Symbol        Event_Info  \n",
       "0                NaN               NaN               NaN  \n",
       "1                NaN               NaN               NaN  \n",
       "2                NaN               NaN               NaN  \n",
       "3                NaN               NaN               NaN  \n",
       "4                NaN               NaN               NaN  \n",
       "5                NaN               NaN               NaN  \n",
       "6                NaN               NaN               NaN  \n",
       "7                NaN               NaN               NaN  \n",
       "8                NaN               NaN               NaN  \n",
       "9                NaN               NaN               NaN  \n",
       "10               NaN               NaN               NaN  \n",
       "11               NaN               NaN               NaN  \n",
       "12               NaN               NaN               NaN  \n",
       "13               NaN               NaN               NaN  \n",
       "14               NaN               NaN               NaN  \n",
       "15               NaN               NaN               NaN  \n",
       "16               NaN               NaN               NaN  \n",
       "17               NaN               NaN               NaN  \n",
       "18               NaN               NaN               NaN  \n",
       "19             MAGI2             MAGI2  MAGI2-Intragenic  \n",
       "\n",
       "[20 rows x 23 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>Cancer Type</th>\n",
       "      <th>Cancer Type Detailed</th>\n",
       "      <th>Tumor Stage</th>\n",
       "      <th>Sample Type</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Diagnosis Age</th>\n",
       "      <th>Smoke Status</th>\n",
       "      <th>TMB (nonsynonymous)</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>...</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>SNP_event</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Codons</th>\n",
       "      <th>Exon_Number</th>\n",
       "      <th>VAR_TYPE_SX</th>\n",
       "      <th>Site1_Hugo_Symbol</th>\n",
       "      <th>Site2_Hugo_Symbol</th>\n",
       "      <th>Event_Info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>IV</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>KMT2C</td>\n",
       "      <td>...</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>4822.0</td>\n",
       "      <td>cGt/cAt</td>\n",
       "      <td>57/59</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>IV</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>...</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>12.0</td>\n",
       "      <td>Ggt/Tgt</td>\n",
       "      <td>5-Feb</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>IV</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>SOX9</td>\n",
       "      <td>...</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>DEL</td>\n",
       "      <td>C&gt;-</td>\n",
       "      <td>236.0</td>\n",
       "      <td>aCc/ac</td>\n",
       "      <td>3-Mar</td>\n",
       "      <td>Truncation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>IV</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>APC</td>\n",
       "      <td>...</td>\n",
       "      <td>stop_gained,splice_region_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>216.0</td>\n",
       "      <td>Cga/Tga</td>\n",
       "      <td>16-Jul</td>\n",
       "      <td>Truncation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>IV</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>APC</td>\n",
       "      <td>...</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G&gt;T</td>\n",
       "      <td>1286.0</td>\n",
       "      <td>Gaa/Taa</td>\n",
       "      <td>16/16</td>\n",
       "      <td>Truncation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>IV</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>QKI</td>\n",
       "      <td>...</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>47.0</td>\n",
       "      <td>Gaa/Aaa</td>\n",
       "      <td>8-Jan</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>IV</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>FBXW7</td>\n",
       "      <td>...</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>505.0</td>\n",
       "      <td>Cgc/Tgc</td>\n",
       "      <td>12-Oct</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>IV</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>PTPN11</td>\n",
       "      <td>...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>DEL</td>\n",
       "      <td>TTTC&gt;-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>IV</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>GLI1</td>\n",
       "      <td>...</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G&gt;T</td>\n",
       "      <td>488.0</td>\n",
       "      <td>Gct/Tct</td>\n",
       "      <td>12-Nov</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Patient0001</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>IV</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Female</td>\n",
       "      <td>67</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.333333333</td>\n",
       "      <td>ATM</td>\n",
       "      <td>...</td>\n",
       "      <td>splice_region_variant,intron_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>1026.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>II</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>SOX9</td>\n",
       "      <td>...</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>DEL</td>\n",
       "      <td>G&gt;-</td>\n",
       "      <td>389.0</td>\n",
       "      <td>ccG/cc</td>\n",
       "      <td>3-Mar</td>\n",
       "      <td>Truncation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>II</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>EPHB1</td>\n",
       "      <td>...</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G&gt;A</td>\n",
       "      <td>743.0</td>\n",
       "      <td>cGg/cAg</td>\n",
       "      <td>16-Dec</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>II</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>FBXW7</td>\n",
       "      <td>...</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T&gt;C</td>\n",
       "      <td>560.0</td>\n",
       "      <td>gAt/gGt</td>\n",
       "      <td>12-Nov</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>II</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>APC</td>\n",
       "      <td>...</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>DEL</td>\n",
       "      <td>C&gt;-</td>\n",
       "      <td>1440.0</td>\n",
       "      <td>Cca/ca</td>\n",
       "      <td>16/16</td>\n",
       "      <td>Truncation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>II</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>RARA</td>\n",
       "      <td>...</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>DEL</td>\n",
       "      <td>AAG&gt;-</td>\n",
       "      <td>164.0</td>\n",
       "      <td>AAG/-</td>\n",
       "      <td>9-May</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>II</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>FBXW7</td>\n",
       "      <td>...</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>465.0</td>\n",
       "      <td>cGt/cAt</td>\n",
       "      <td>12-Sep</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>II</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>...</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;A</td>\n",
       "      <td>12.0</td>\n",
       "      <td>gGt/gTt</td>\n",
       "      <td>5-Feb</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>II</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>TP53</td>\n",
       "      <td>...</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>175.0</td>\n",
       "      <td>cGc/cAc</td>\n",
       "      <td>11-May</td>\n",
       "      <td>Substitution/Indel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Patient0002</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>II</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Male</td>\n",
       "      <td>75</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.3</td>\n",
       "      <td>APC</td>\n",
       "      <td>...</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>213.0</td>\n",
       "      <td>Cga/Tga</td>\n",
       "      <td>16-Jun</td>\n",
       "      <td>Truncation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Patient0003</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>III</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Female</td>\n",
       "      <td>45</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0.5</td>\n",
       "      <td>APC</td>\n",
       "      <td>...</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>1450.0</td>\n",
       "      <td>Cga/Tga</td>\n",
       "      <td>16/16</td>\n",
       "      <td>Truncation</td>\n",
       "      <td>MAGI2</td>\n",
       "      <td>MAGI2</td>\n",
       "      <td>MAGI2-Intragenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows × 23 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 13
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-05T13:53:06.706235Z",
     "start_time": "2025-05-05T13:53:06.703946Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Function to handle the conversion\n",
    "def convert_exon_number(val):\n",
    "    try:\n",
    "        # First, try to convert to 'Month-Year' format (e.g., 'Sep-89' -> '09/89')\n",
    "        return pd.to_datetime(val, format='%b-%y').strftime('%m/%y')\n",
    "    except ValueError:\n",
    "        pass\n",
    "\n",
    "    try:\n",
    "        # Then, try to convert to 'DD-Mon' format (e.g., '14-Sep' -> '09/14')\n",
    "        date_obj = pd.to_datetime(val, format='%d-%b', errors='raise')\n",
    "        return date_obj.strftime('%m/%d')\n",
    "    except ValueError:\n",
    "        # If neither format matches, return the value as is (non-date-like string)\n",
    "        return val"
   ],
   "outputs": [],
   "execution_count": 14
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-05T13:53:10.035114Z",
     "start_time": "2025-05-05T13:53:07.473959Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Apply the function to the column\n",
    "data_for_model.loc[:, 'Exon_Number'] = data_for_model['Exon_Number'].apply(convert_exon_number)"
   ],
   "outputs": [],
   "execution_count": 15
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-05T13:53:10.548614Z",
     "start_time": "2025-05-05T13:53:10.543165Z"
    }
   },
   "cell_type": "code",
   "source": "data_for_model[\"Exon_Number\"].isnull().sum()",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(15278)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 16
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-05T13:53:16.122436Z",
     "start_time": "2025-05-05T13:53:16.116836Z"
    }
   },
   "cell_type": "code",
   "source": "data_for_model[\"Cancer Type\"].value_counts()",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Cancer Type\n",
       "Colorectal Carcinoma                     28396\n",
       "Non Small Cell Lung Cancer               19526\n",
       "Gastric Cancer                           10832\n",
       "Liver Hepatocellular Carcinoma            9633\n",
       "Esophageal Carcinoma                      6298\n",
       "Intrahepatic Cholangiocarcinoma           3939\n",
       "Small Cell Lung Cancer                    3025\n",
       "Pancreatic Cancer                         2957\n",
       "Extrahepatic Cholangiocarcinoma           2883\n",
       "Breast Carcinoma                          2350\n",
       "Soft Tissue Sarcoma                       2270\n",
       "Gallbladder Carcinoma                     2192\n",
       "Ovarian Carcinoma                         1851\n",
       "Urothelial Carcinoma                      1631\n",
       "Kidney Renal Cell Carcinoma               1500\n",
       "Uterine Corpus Endometrial Carcinoma      1358\n",
       "Cancer of Unknown Primary                 1068\n",
       "Head and Neck Carcinoma                    997\n",
       "Carcinoma of Uterine Cervix                787\n",
       "Small Bowel Carcinoma                      755\n",
       "Bone Sarcoma                               601\n",
       "Gastrointestinal Neuroendocrine Tumor      415\n",
       "Melanoma                                   388\n",
       "Thymic Tumor                               170\n",
       "Thyroid Carcinoma                           84\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 17
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-05T13:53:17.443715Z",
     "start_time": "2025-05-05T13:53:17.441337Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def create_age_range(x):\n",
    "    if x <= 10:\n",
    "        return \"0-10\"\n",
    "    elif x <= 20:\n",
    "        return \"11-20\"\n",
    "    elif x <= 30:\n",
    "        return \"21-30\"\n",
    "    elif x <= 40:\n",
    "        return \"31-40\"\n",
    "    elif x <= 50:\n",
    "        return \"41-50\"\n",
    "    elif x <= 60:\n",
    "        return \"51-60\"\n",
    "    elif x <= 70:\n",
    "        return \"61-70\"\n",
    "    elif x <= 80:\n",
    "        return \"71-80\"\n",
    "    else:\n",
    "        return \"80+\""
   ],
   "outputs": [],
   "execution_count": 18
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-05T13:53:24.848236Z",
     "start_time": "2025-05-05T13:53:24.825699Z"
    }
   },
   "cell_type": "code",
   "source": "data_for_model.loc[:, 'Diagnosis Age'] = data_for_model['Diagnosis Age'].astype(int).apply(create_age_range).astype(\"category\")",
   "outputs": [],
   "execution_count": 19
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-05T13:53:33.443324Z",
     "start_time": "2025-05-05T13:53:33.433915Z"
    }
   },
   "cell_type": "code",
   "source": [
    "tt = dict(enumerate(data_for_model[\"Exon_Number\"].astype('category').cat.categories))\n",
    "tt.values()\n",
    "# data[object_columns] = data[object_columns].astype('category')"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_values(['01/01', '01/02', '01/03', '01/04', '01/05', '01/06', '01/07', '01/08', '01/09', '01/10', '01/11', '01/12', '01/13', '01/14', '01/15', '01/16', '01/17', '01/18', '01/19', '01/20', '01/21', '01/22', '01/23', '01/24', '01/25', '01/26', '01/27', '01/28', '01/29', '01/30', '01/31', '01/33', '01/34', '01/35', '01/36', '01/37', '01/38', '01/40', '01/43', '01/45', '01/47', '01/49', '01/51', '01/52', '01/54', '01/57', '01/59', '01/79', '01/87', '01/89', '01/91', '02/02', '02/03', '02/04', '02/05', '02/06', '02/07', '02/08', '02/09', '02/10', '02/11', '02/12', '02/13', '02/14', '02/15', '02/16', '02/17', '02/18', '02/19', '02/20', '02/21', '02/22', '02/23', '02/24', '02/25', '02/26', '02/27', '02/28', '02/29', '02/30', '02/31', '02/32', '02/33', '02/34', '02/35', '02/36', '02/37', '02/38', '02/39', '02/40', '02/42', '02/43', '02/44', '02/45', '02/46', '02/49', '02/51', '02/52', '02/54', '02/57', '02/58', '02/59', '02/63', '02/79', '02/87', '02/89', '02/91', '03/03', '03/04', '03/05', '03/06', '03/07', '03/08', '03/09', '03/10', '03/11', '03/12', '03/13', '03/14', '03/15', '03/16', '03/17', '03/18', '03/19', '03/20', '03/21', '03/22', '03/23', '03/24', '03/25', '03/26', '03/27', '03/28', '03/29', '03/30', '03/31', '03/32', '03/33', '03/34', '03/35', '03/36', '03/37', '03/38', '03/39', '03/40', '03/42', '03/43', '03/44', '03/45', '03/46', '03/47', '03/49', '03/51', '03/52', '03/54', '03/57', '03/58', '03/59', '03/63', '03/79', '03/87', '03/89', '03/91', '04/04', '04/05', '04/06', '04/07', '04/08', '04/09', '04/10', '04/11', '04/12', '04/13', '04/14', '04/15', '04/16', '04/17', '04/18', '04/19', '04/20', '04/21', '04/22', '04/23', '04/24', '04/25', '04/26', '04/27', '04/28', '04/29', '04/30', '04/31', '04/32', '04/33', '04/34', '04/35', '04/36', '04/37', '04/38', '04/39', '04/40', '04/42', '04/43', '04/44', '04/45', '04/46', '04/47', '04/49', '04/51', '04/52', '04/54', '04/57', '04/58', '04/59', '04/63', '04/79', '04/87', '04/89', '04/91', '05/05', '05/06', '05/07', '05/08', '05/09', '05/10', '05/11', '05/12', '05/13', '05/14', '05/15', '05/16', '05/17', '05/18', '05/19', '05/20', '05/21', '05/22', '05/23', '05/24', '05/25', '05/26', '05/27', '05/28', '05/29', '05/30', '05/31', '05/32', '05/33', '05/34', '05/35', '05/36', '05/37', '05/38', '05/39', '05/40', '05/42', '05/43', '05/44', '05/45', '05/46', '05/47', '05/49', '05/51', '05/52', '05/54', '05/57', '05/58', '05/59', '05/63', '05/79', '05/87', '05/89', '05/91', '06/06', '06/07', '06/08', '06/09', '06/10', '06/11', '06/12', '06/13', '06/14', '06/15', '06/16', '06/17', '06/18', '06/19', '06/20', '06/21', '06/22', '06/23', '06/24', '06/25', '06/26', '06/27', '06/28', '06/29', '06/30', '06/31', '06/32', '06/33', '06/34', '06/35', '06/36', '06/37', '06/38', '06/39', '06/40', '06/42', '06/43', '06/45', '06/46', '06/47', '06/49', '06/51', '06/52', '06/54', '06/57', '06/58', '06/59', '06/63', '06/79', '06/87', '06/89', '06/91', '07/07', '07/08', '07/09', '07/10', '07/11', '07/12', '07/13', '07/14', '07/15', '07/16', '07/17', '07/18', '07/19', '07/20', '07/21', '07/22', '07/23', '07/24', '07/25', '07/26', '07/27', '07/28', '07/29', '07/30', '07/31', '07/32', '07/33', '07/34', '07/35', '07/36', '07/37', '07/38', '07/39', '07/40', '07/42', '07/43', '07/44', '07/45', '07/46', '07/47', '07/49', '07/52', '07/54', '07/57', '07/58', '07/59', '07/63', '07/79', '07/87', '07/89', '07/91', '08/08', '08/09', '08/10', '08/11', '08/12', '08/13', '08/14', '08/15', '08/16', '08/17', '08/18', '08/19', '08/20', '08/21', '08/22', '08/23', '08/24', '08/25', '08/26', '08/27', '08/28', '08/29', '08/30', '08/31', '08/32', '08/33', '08/34', '08/35', '08/36', '08/37', '08/38', '08/39', '08/40', '08/42', '08/43', '08/45', '08/46', '08/47', '08/49', '08/52', '08/54', '08/57', '08/58', '08/59', '08/63', '08/79', '08/87', '08/89', '08/91', '09/09', '09/10', '09/11', '09/12', '09/13', '09/14', '09/15', '09/16', '09/17', '09/18', '09/19', '09/20', '09/21', '09/22', '09/23', '09/24', '09/25', '09/26', '09/27', '09/28', '09/29', '09/30', '09/31', '09/32', '09/33', '09/34', '09/35', '09/36', '09/37', '09/38', '09/39', '09/40', '09/42', '09/43', '09/44', '09/45', '09/46', '09/47', '09/49', '09/51', '09/52', '09/54', '09/57', '09/58', '09/59', '09/63', '09/79', '09/87', '09/89', '09/91', '10/10', '10/11', '10/12', '10/13', '10/14', '10/15', '10/16', '10/17', '10/18', '10/19', '10/20', '10/21', '10/22', '10/23', '10/24', '10/25', '10/26', '10/27', '10/28', '10/29', '10/30', '10/31', '10/32', '10/33', '10/34', '10/35', '10/36', '10/37', '10/38', '10/39', '10/40', '10/42', '10/43', '10/44', '10/45', '10/46', '10/47', '10/49', '10/51', '10/52', '10/54', '10/57', '10/58', '10/59', '10/63', '10/79', '10/87', '10/89', '10/91', '11/11', '11/12', '11/13', '11/14', '11/15', '11/16', '11/17', '11/18', '11/19', '11/20', '11/21', '11/22', '11/23', '11/24', '11/25', '11/26', '11/27', '11/28', '11/29', '11/30', '11/31', '11/32', '11/33', '11/34', '11/35', '11/36', '11/37', '11/38', '11/39', '11/40', '11/42', '11/43', '11/44', '11/45', '11/46', '11/47', '11/51', '11/52', '11/54', '11/57', '11/58', '11/59', '11/63', '11/79', '11/87', '11/89', '11/91', '12/12', '12/13', '12/14', '12/15', '12/16', '12/17', '12/18', '12/19', '12/20', '12/21', '12/22', '12/23', '12/24', '12/25', '12/26', '12/27', '12/28', '12/29', '12/30', '12/31', '12/32', '12/33', '12/34', '12/35', '12/36', '12/37', '12/38', '12/39', '12/40', '12/42', '12/43', '12/44', '12/45', '12/46', '12/47', '12/49', '12/51', '12/52', '12/54', '12/57', '12/58', '12/59', '12/63', '12/79', '12/87', '12/89', '12/91', '13/13', '13/14', '13/15', '13/16', '13/17', '13/18', '13/19', '13/20', '13/21', '13/22', '13/23', '13/24', '13/25', '13/26', '13/27', '13/28', '13/29', '13/30', '13/31', '13/32', '13/33', '13/34', '13/35', '13/36', '13/37', '13/38', '13/39', '13/40', '13/42', '13/43', '13/44', '13/45', '13/46', '13/47', '13/49', '13/51', '13/52', '13/54', '13/57', '13/58', '13/59', '13/63', '13/79', '13/87', '13/89', '13/91', '14/14', '14/15', '14/16', '14/17', '14/18', '14/19', '14/20', '14/21', '14/22', '14/23', '14/24', '14/25', '14/26', '14/27', '14/28', '14/29', '14/30', '14/31', '14/32', '14/33', '14/34', '14/35', '14/36', '14/37', '14/38', '14/39', '14/40', '14/42', '14/43', '14/44', '14/45', '14/46', '14/47', '14/49', '14/51', '14/52', '14/54', '14/57', '14/58', '14/59', '14/63', '14/79', '14/89', '14/91', '15/15', '15/16', '15/17', '15/18', '15/19', '15/20', '15/21', '15/22', '15/23', '15/24', '15/25', '15/26', '15/27', '15/28', '15/29', '15/30', '15/31', '15/32', '15/33', '15/34', '15/35', '15/36', '15/37', '15/38', '15/39', '15/40', '15/42', '15/43', '15/44', '15/45', '15/46', '15/47', '15/49', '15/51', '15/52', '15/54', '15/57', '15/58', '15/59', '15/63', '15/79', '15/87', '15/89', '15/91', '16/16', '16/17', '16/18', '16/19', '16/20', '16/21', '16/22', '16/23', '16/24', '16/25', '16/26', '16/27', '16/28', '16/29', '16/30', '16/31', '16/32', '16/33', '16/34', '16/35', '16/36', '16/37', '16/38', '16/39', '16/40', '16/42', '16/43', '16/44', '16/45', '16/46', '16/47', '16/49', '16/51', '16/52', '16/54', '16/57', '16/58', '16/59', '16/63', '16/79', '16/87', '16/89', '16/91', '17/17', '17/18', '17/19', '17/20', '17/21', '17/22', '17/23', '17/24', '17/25', '17/26', '17/27', '17/28', '17/29', '17/30', '17/31', '17/32', '17/33', '17/34', '17/35', '17/36', '17/37', '17/38', '17/39', '17/40', '17/42', '17/43', '17/44', '17/46', '17/47', '17/49', '17/51', '17/52', '17/54', '17/57', '17/58', '17/59', '17/63', '17/79', '17/87', '17/89', '17/91', '18/18', '18/19', '18/20', '18/21', '18/22', '18/23', '18/24', '18/25', '18/26', '18/27', '18/28', '18/29', '18/30', '18/31', '18/32', '18/33', '18/34', '18/35', '18/36', '18/37', '18/38', '18/39', '18/40', '18/42', '18/43', '18/44', '18/45', '18/46', '18/47', '18/49', '18/51', '18/52', '18/54', '18/57', '18/58', '18/59', '18/63', '18/79', '18/87', '18/89', '18/91', '19/19', '19/20', '19/21', '19/22', '19/23', '19/24', '19/25', '19/26', '19/27', '19/28', '19/29', '19/30', '19/31', '19/32', '19/33', '19/34', '19/35', '19/36', '19/37', '19/38', '19/39', '19/40', '19/42', '19/43', '19/44', '19/45', '19/46', '19/47', '19/49', '19/51', '19/52', '19/54', '19/57', '19/58', '19/59', '19/63', '19/79', '19/87', '19/89', '19/91', '20/20', '20/21', '20/22', '20/23', '20/24', '20/25', '20/26', '20/27', '20/28', '20/29', '20/30', '20/31', '20/32', '20/33', '20/34', '20/35', '20/36', '20/37', '20/38', '20/39', '20/40', '20/42', '20/43', '20/44', '20/45', '20/46', '20/47', '20/49', '20/51', '20/52', '20/54', '20/57', '20/58', '20/59', '20/63', '20/79', '20/87', '20/89', '20/91', '21/21', '21/22', '21/23', '21/24', '21/25', '21/26', '21/27', '21/28', '21/29', '21/30', '21/31', '21/32', '21/33', '21/34', '21/35', '21/36', '21/37', '21/38', '21/39', '21/40', '21/42', '21/43', '21/44', '21/45', '21/46', '21/47', '21/49', '21/51', '21/52', '21/54', '21/57', '21/58', '21/59', '21/63', '21/79', '21/87', '21/89', '21/91', '22/22', '22/23', '22/24', '22/25', '22/26', '22/27', '22/28', '22/29', '22/30', '22/31', '22/32', '22/33', '22/34', '22/35', '22/36', '22/37', '22/38', '22/39', '22/40', '22/42', '22/43', '22/44', '22/45', '22/46', '22/47', '22/49', '22/51', '22/52', '22/54', '22/57', '22/59', '22/63', '22/79', '22/87', '22/89', '22/91', '23/23', '23/24', '23/25', '23/26', '23/27', '23/28', '23/29', '23/30', '23/31', '23/32', '23/33', '23/34', '23/35', '23/36', '23/37', '23/38', '23/39', '23/40', '23/42', '23/43', '23/44', '23/45', '23/46', '23/47', '23/49', '23/51', '23/52', '23/54', '23/57', '23/58', '23/59', '23/63', '23/79', '23/87', '23/89', '23/91', '24/24', '24/25', '24/26', '24/27', '24/28', '24/29', '24/30', '24/31', '24/32', '24/33', '24/34', '24/35', '24/36', '24/37', '24/38', '24/39', '24/40', '24/42', '24/43', '24/45', '24/46', '24/47', '24/49', '24/51', '24/52', '24/54', '24/57', '24/58', '24/59', '24/63', '24/79', '24/87', '24/89', '24/91', '25/25', '25/26', '25/27', '25/28', '25/29', '25/30', '25/31', '25/32', '25/33', '25/34', '25/35', '25/36', '25/37', '25/38', '25/39', '25/40', '25/43', '25/44', '25/45', '25/46', '25/47', '25/49', '25/51', '25/52', '25/54', '25/57', '25/58', '25/59', '25/63', '25/79', '25/87', '25/89', '25/91', '26/26', '26/27', '26/28', '26/29', '26/30', '26/31', '26/32', '26/33', '26/34', '26/35', '26/36', '26/37', '26/38', '26/39', '26/40', '26/42', '26/43', '26/44', '26/45', '26/46', '26/47', '26/49', '26/51', '26/52', '26/54', '26/57', '26/58', '26/59', '26/63', '26/79', '26/87', '26/89', '26/91', '27/27', '27/28', '27/29', '27/30', '27/31', '27/32', '27/33', '27/34', '27/35', '27/36', '27/37', '27/38', '27/39', '27/40', '27/42', '27/43', '27/44', '27/45', '27/46', '27/47', '27/49', '27/51', '27/52', '27/54', '27/57', '27/59', '27/63', '27/79', '27/87', '27/89', '27/91', '28/28', '28/29', '28/30', '28/31', '28/32', '28/33', '28/34', '28/35', '28/36', '28/37', '28/38', '28/39', '28/40', '28/42', '28/43', '28/44', '28/45', '28/46', '28/47', '28/49', '28/51', '28/52', '28/54', '28/57', '28/58', '28/59', '28/63', '28/79', '28/89', '28/91', '29/29', '29/30', '29/31', '29/32', '29/33', '29/34', '29/35', '29/36', '29/37', '29/38', '29/39', '29/40', '29/42', '29/43', '29/44', '29/45', '29/46', '29/47', '29/49', '29/52', '29/54', '29/57', '29/58', '29/59', '29/63', '29/79', '29/87', '29/89', '29/91', '30/30', '30/31', '30/32', '30/33', '30/34', '30/35', '30/36', '30/37', '30/38', '30/39', '30/40', '30/42', '30/43', '30/44', '30/45', '30/46', '30/47', '30/49', '30/51', '30/52', '30/54', '30/57', '30/58', '30/63', '30/79', '30/87', '30/89', '30/91', '31/31', '31/32', '31/33', '31/34', '31/35', '31/36', '31/37', '31/38', '31/39', '31/40', '31/42', '31/43', '31/44', '31/45', '31/46', '31/47', '31/49', '31/51', '31/52', '31/54', '31/57', '31/58', '31/59', '31/63', '31/79', '31/89', '31/91', '32/32', '32/33', '32/34', '32/35', '32/36', '32/37', '32/38', '32/39', '32/40', '32/42', '32/43', '32/44', '32/45', '32/46', '32/47', '32/49', '32/51', '32/52', '32/54', '32/57', '32/58', '32/59', '32/63', '32/79', '32/87', '32/89', '32/91', '33/33', '33/34', '33/35', '33/36', '33/37', '33/38', '33/39', '33/40', '33/42', '33/43', '33/44', '33/45', '33/46', '33/47', '33/49', '33/51', '33/52', '33/54', '33/57', '33/58', '33/59', '33/63', '33/79', '33/87', '33/89', '33/91', '34/34', '34/35', '34/36', '34/37', '34/38', '34/39', '34/40', '34/42', '34/43', '34/44', '34/45', '34/46', '34/47', '34/49', '34/51', '34/52', '34/54', '34/57', '34/58', '34/59', '34/63', '34/79', '34/87', '34/89', '34/91', '35/35', '35/36', '35/37', '35/38', '35/39', '35/40', '35/42', '35/43', '35/44', '35/45', '35/46', '35/47', '35/49', '35/51', '35/52', '35/54', '35/57', '35/58', '35/59', '35/63', '35/79', '35/87', '35/89', '35/91', '36/36', '36/37', '36/38', '36/39', '36/40', '36/42', '36/43', '36/44', '36/45', '36/46', '36/47', '36/49', '36/51', '36/52', '36/54', '36/57', '36/58', '36/59', '36/63', '36/79', '36/87', '36/89', '36/91', '37/37', '37/38', '37/39', '37/40', '37/42', '37/43', '37/44', '37/45', '37/46', '37/47', '37/49', '37/51', '37/52', '37/54', '37/57', '37/58', '37/59', '37/63', '37/79', '37/87', '37/89', '37/91', '38/38', '38/39', '38/40', '38/42', '38/43', '38/44', '38/45', '38/46', '38/47', '38/49', '38/51', '38/52', '38/54', '38/57', '38/58', '38/59', '38/63', '38/79', '38/87', '38/89', '38/91', '39/39', '39/40', '39/42', '39/43', '39/44', '39/45', '39/46', '39/47', '39/49', '39/51', '39/52', '39/54', '39/57', '39/58', '39/59', '39/63', '39/79', '39/87', '39/89', '39/91', '40/40', '40/42', '40/43', '40/44', '40/45', '40/46', '40/47', '40/49', '40/51', '40/52', '40/54', '40/57', '40/58', '40/63', '40/79', '40/87', '40/89', '40/91', '41/42', '41/43', '41/44', '41/45', '41/46', '41/47', '41/49', '41/51', '41/52', '41/54', '41/57', '41/58', '41/59', '41/63', '41/79', '41/87', '41/89', '41/91', '42/42', '42/43', '42/44', '42/45', '42/46', '42/47', '42/49', '42/51', '42/52', '42/54', '42/57', '42/58', '42/59', '42/63', '42/79', '42/87', '42/89', '42/91', '43/43', '43/44', '43/45', '43/46', '43/47', '43/49', '43/51', '43/52', '43/54', '43/57', '43/58', '43/59', '43/63', '43/79', '43/87', '43/89', '43/91', '44/44', '44/45', '44/46', '44/47', '44/49', '44/51', '44/52', '44/54', '44/57', '44/58', '44/59', '44/63', '44/79', '44/87', '44/89', '44/91', '45/45', '45/46', '45/47', '45/49', '45/51', '45/52', '45/54', '45/57', '45/58', '45/59', '45/63', '45/79', '45/87', '45/89', '45/91', '46/46', '46/47', '46/49', '46/51', '46/52', '46/54', '46/57', '46/58', '46/59', '46/63', '46/79', '46/87', '46/89', '46/91', '47/47', '47/49', '47/51', '47/52', '47/54', '47/57', '47/58', '47/59', '47/63', '47/79', '47/87', '47/89', '47/91', '48/49', '48/51', '48/52', '48/54', '48/57', '48/58', '48/59', '48/63', '48/79', '48/87', '48/89', '48/91', '49/49', '49/51', '49/52', '49/54', '49/57', '49/58', '49/59', '49/63', '49/79', '49/87', '49/89', '49/91', '50/51', '50/52', '50/54', '50/57', '50/58', '50/59', '50/63', '50/79', '50/87', '50/89', '50/91', '51/51', '51/52', '51/54', '51/57', '51/58', '51/59', '51/63', '51/79', '51/87', '51/89', '51/91', '52/52', '52/54', '52/57', '52/58', '52/59', '52/63', '52/79', '52/87', '52/89', '52/91', '53/54', '53/57', '53/58', '53/59', '53/63', '53/79', '53/87', '53/89', '53/91', '54/54', '54/57', '54/59', '54/63', '54/79', '54/87', '54/89', '54/91', '55/57', '55/58', '55/59', '55/63', '55/79', '55/87', '55/89', '55/91', '56/57', '56/58', '56/59', '56/63', '56/79', '56/87', '56/89', '56/91', '57/57', '57/58', '57/59', '57/63', '57/79', '57/87', '57/89', '57/91', '58/59', '58/63', '58/79', '58/87', '58/89', '58/91', '59/59', '59/63', '59/79', '59/87', '59/89', '59/91', '60/63', '60/79', '60/87', '60/89', '60/91', '61/79', '61/87', '61/89', '61/91', '62/63', '62/79', '62/87', '62/89', '62/91', '63/63', '63/79', '63/87', '63/89', '63/91', '64/79', '64/87', '64/89', '64/91', '65/79', '65/87', '65/89', '65/91', '66/79', '66/87', '66/89', '66/91', '67/79', '67/87', '67/89', '67/91', '68/79', '68/87', '68/89', '68/91', '69/79', '69/87', '69/89', '69/91', '70/79', '70/87', '70/89', '70/91', '71/79', '71/87', '71/89', '71/91', '72/79', '72/87', '72/89', '72/91', '73/79', '73/87', '73/89', '73/91', '74/79', '74/87', '74/89', '74/91', '75/79', '75/87', '75/89', '75/91', '76/79', '76/87', '76/89', '76/91', '77/79', '77/87', '77/89', '77/91', '78/79', '78/87', '78/89', '78/91', '79/79', '79/87', '79/89', '79/91', '80/87', '80/89', '80/91', '81/89', '81/91', '82/87', '82/89', '82/91', '83/87', '83/89', '83/91', '84/87', '84/89', '84/91', '85/87', '85/89', '85/91', '86/87', '86/89', '86/91', '87/87', '87/89', '87/91', '88/89', '88/91', '89/89', '89/91', '90/91', '91/91'])"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 21
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-05T13:53:38.303410Z",
     "start_time": "2025-05-05T13:53:38.266341Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Filter cancer types with at least 2000 samples\n",
    "cancer_counts = data_for_model['Cancer Type'].value_counts()\n",
    "valid_cancer_types = cancer_counts[cancer_counts >= 2000].index\n",
    "data_for_model = data_for_model[data_for_model['Cancer Type'].isin(valid_cancer_types)]\n",
    "data_for_model = data_for_model[data_for_model['Chromosome'].notnull()]"
   ],
   "outputs": [],
   "execution_count": 22
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "data_for_lift = data_for_model.copy()",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "data_for_lift[\"Consequence\"].unique()",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-05T13:53:48.169290Z",
     "start_time": "2025-05-05T13:53:47.724034Z"
    }
   },
   "cell_type": "code",
   "source": [
    "data_for_model['Consequence'].str.split(',')\n",
    "dummy_vars = data_for_model['Consequence'].str.split(',').explode().str.get_dummies().groupby(level=0).sum()\n",
    "data_for_model = data_for_model.join(dummy_vars)\n",
    "data_for_model.drop('Consequence', axis=1, inplace=True)"
   ],
   "outputs": [],
   "execution_count": 23
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-05T13:53:48.823961Z",
     "start_time": "2025-05-05T13:53:48.815002Z"
    }
   },
   "cell_type": "code",
   "source": "data_for_model[data_for_model[\"PATIENT_ID\"] == \"Patient8178\"]",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "        PATIENT_ID           Cancer Type       Cancer Type Detailed  \\\n",
       "87594  Patient8178  Colorectal Carcinoma  Colorectal Adenocarcinoma   \n",
       "87595  Patient8178  Colorectal Carcinoma  Colorectal Adenocarcinoma   \n",
       "87596  Patient8178  Colorectal Carcinoma  Colorectal Adenocarcinoma   \n",
       "87597  Patient8178  Colorectal Carcinoma  Colorectal Adenocarcinoma   \n",
       "87598  Patient8178  Colorectal Carcinoma  Colorectal Adenocarcinoma   \n",
       "...            ...                   ...                        ...   \n",
       "87797  Patient8178  Colorectal Carcinoma  Colorectal Adenocarcinoma   \n",
       "87798  Patient8178  Colorectal Carcinoma  Colorectal Adenocarcinoma   \n",
       "87799  Patient8178  Colorectal Carcinoma  Colorectal Adenocarcinoma   \n",
       "87800  Patient8178  Colorectal Carcinoma  Colorectal Adenocarcinoma   \n",
       "87801  Patient8178  Colorectal Carcinoma  Colorectal Adenocarcinoma   \n",
       "\n",
       "      Tumor Stage Sample Type     Sex Diagnosis Age Smoke Status  \\\n",
       "87594           I     Primary  Female         71-80      Unknown   \n",
       "87595           I     Primary  Female         71-80      Unknown   \n",
       "87596           I     Primary  Female         71-80      Unknown   \n",
       "87597           I     Primary  Female         71-80      Unknown   \n",
       "87598           I     Primary  Female         71-80      Unknown   \n",
       "...           ...         ...     ...           ...          ...   \n",
       "87797           I     Primary  Female         71-80      Unknown   \n",
       "87798           I     Primary  Female         71-80      Unknown   \n",
       "87799           I     Primary  Female         71-80      Unknown   \n",
       "87800           I     Primary  Female         71-80      Unknown   \n",
       "87801           I     Primary  Female         71-80      Unknown   \n",
       "\n",
       "      TMB (nonsynonymous) Hugo_Symbol  ... splice_acceptor_variant  \\\n",
       "87594         6.166666667      TGFBR2  ...                       0   \n",
       "87595         6.166666667        POLE  ...                       0   \n",
       "87596         6.166666667      CTNNA1  ...                       0   \n",
       "87597         6.166666667      PTPN11  ...                       0   \n",
       "87598         6.166666667       SETD2  ...                       0   \n",
       "...                   ...         ...  ...                     ...   \n",
       "87797         6.166666667        BRAF  ...                       0   \n",
       "87798         6.166666667     PIK3C2B  ...                       0   \n",
       "87799         6.166666667       RECQL  ...                       0   \n",
       "87800         6.166666667        NSD1  ...                       0   \n",
       "87801         6.166666667       PARP1  ...                       0   \n",
       "\n",
       "       splice_donor_variant  splice_region_variant start_lost  \\\n",
       "87594                     0                      0          0   \n",
       "87595                     0                      0          0   \n",
       "87596                     0                      0          0   \n",
       "87597                     0                      0          0   \n",
       "87598                     0                      0          0   \n",
       "...                     ...                    ...        ...   \n",
       "87797                     0                      0          0   \n",
       "87798                     0                      0          0   \n",
       "87799                     0                      0          0   \n",
       "87800                     0                      0          0   \n",
       "87801                     0                      0          0   \n",
       "\n",
       "      start_retained_variant  stop_gained stop_lost stop_retained_variant  \\\n",
       "87594                      0            0         0                     0   \n",
       "87595                      0            0         0                     0   \n",
       "87596                      0            0         0                     0   \n",
       "87597                      0            0         0                     0   \n",
       "87598                      0            0         0                     0   \n",
       "...                      ...          ...       ...                   ...   \n",
       "87797                      0            0         0                     0   \n",
       "87798                      0            0         0                     0   \n",
       "87799                      0            0         0                     0   \n",
       "87800                      0            0         0                     0   \n",
       "87801                      0            0         0                     0   \n",
       "\n",
       "      synonymous_variant upstream_gene_variant  \n",
       "87594                  0                     0  \n",
       "87595                  0                     0  \n",
       "87596                  0                     0  \n",
       "87597                  0                     0  \n",
       "87598                  0                     0  \n",
       "...                  ...                   ...  \n",
       "87797                  0                     0  \n",
       "87798                  0                     0  \n",
       "87799                  0                     0  \n",
       "87800                  0                     0  \n",
       "87801                  0                     0  \n",
       "\n",
       "[208 rows x 47 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>Cancer Type</th>\n",
       "      <th>Cancer Type Detailed</th>\n",
       "      <th>Tumor Stage</th>\n",
       "      <th>Sample Type</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Diagnosis Age</th>\n",
       "      <th>Smoke Status</th>\n",
       "      <th>TMB (nonsynonymous)</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>...</th>\n",
       "      <th>splice_acceptor_variant</th>\n",
       "      <th>splice_donor_variant</th>\n",
       "      <th>splice_region_variant</th>\n",
       "      <th>start_lost</th>\n",
       "      <th>start_retained_variant</th>\n",
       "      <th>stop_gained</th>\n",
       "      <th>stop_lost</th>\n",
       "      <th>stop_retained_variant</th>\n",
       "      <th>synonymous_variant</th>\n",
       "      <th>upstream_gene_variant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>87594</th>\n",
       "      <td>Patient8178</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>I</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Female</td>\n",
       "      <td>71-80</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>6.166666667</td>\n",
       "      <td>TGFBR2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87595</th>\n",
       "      <td>Patient8178</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>I</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Female</td>\n",
       "      <td>71-80</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>6.166666667</td>\n",
       "      <td>POLE</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87596</th>\n",
       "      <td>Patient8178</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>I</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Female</td>\n",
       "      <td>71-80</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>6.166666667</td>\n",
       "      <td>CTNNA1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87597</th>\n",
       "      <td>Patient8178</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>I</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Female</td>\n",
       "      <td>71-80</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>6.166666667</td>\n",
       "      <td>PTPN11</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87598</th>\n",
       "      <td>Patient8178</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>I</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Female</td>\n",
       "      <td>71-80</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>6.166666667</td>\n",
       "      <td>SETD2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87797</th>\n",
       "      <td>Patient8178</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>I</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Female</td>\n",
       "      <td>71-80</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>6.166666667</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87798</th>\n",
       "      <td>Patient8178</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>I</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Female</td>\n",
       "      <td>71-80</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>6.166666667</td>\n",
       "      <td>PIK3C2B</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87799</th>\n",
       "      <td>Patient8178</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>I</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Female</td>\n",
       "      <td>71-80</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>6.166666667</td>\n",
       "      <td>RECQL</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87800</th>\n",
       "      <td>Patient8178</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>I</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Female</td>\n",
       "      <td>71-80</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>6.166666667</td>\n",
       "      <td>NSD1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87801</th>\n",
       "      <td>Patient8178</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>Colorectal Adenocarcinoma</td>\n",
       "      <td>I</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Female</td>\n",
       "      <td>71-80</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>6.166666667</td>\n",
       "      <td>PARP1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>208 rows × 47 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 24
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "data_for_model.to_csv(\"pan_cancer_data_for_model.csv\", index=False)",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "hypo_data = pd.read_csv(\"hypotheses.csv\")",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "hypo_data_updates = hypo_data[hypo_data[\"support\"] > 2].sort_values([\"cancer_type\", 'support'], ascending=[True, False])\n",
    "hypo_data_updates.head()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "def cancer_type_correlations(df):\n",
    "    \"\"\"\n",
    "    Print cancer type and non-null feature-value pairs for each row in the DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "    - df: DataFrame with columns \"cancer_type\", feature columns, and \"support\".\n",
    "    \"\"\"\n",
    "    corr_list = []\n",
    "    # Iterate through each row\n",
    "    for index, row in df.iterrows():\n",
    "        # Extract cancer type and support\n",
    "        cancer_type = row[\"cancer_type\"]\n",
    "        support = row[\"support\"]\n",
    "\n",
    "        # Get feature-value pairs where the feature value is not null\n",
    "        features = [\n",
    "            f\"{feature}={row[feature]}\"\n",
    "            for feature in df.columns\n",
    "            if feature not in {\"cancer_type\", \"support\"} and not pd.isnull(row[feature])\n",
    "        ]\n",
    "\n",
    "        # Format and print the result\n",
    "        features_str = \", \".join(features)\n",
    "        corr_list.append(f\"{cancer_type}: {features_str}, Support: {support}\")\n",
    "    return corr_list\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "corr_list = cancer_type_correlations(hypo_data_updates)",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "corr_list",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Lift Calculation"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "cancer_type_dummy = data_for_lift['Cancer Type'].str.get_dummies().groupby(level=0).sum()\n",
    "data_for_lift = data_for_lift.join(cancer_type_dummy)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "data_for_lift.head()",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Combine columns to create specific mutation identifiers\n",
    "# data_for_lift['Mutation'] = data_for_lift['Chromosome'] + \"_\" + data_for_lift['Start_Position'].astype(str) + \"_\" + data_for_lift['Variant_Type']\n",
    "data_for_lift['Position'] = data_for_lift['Start_Position'].astype(str) + \"-\" + data_for_lift['End_Position'].astype(str)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "data_for_lift.to_csv(\"data_for_lift.csv\", index=False)",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Select a subset of columns to analyze (e.g., most relevant ones)\n",
    "columns_to_combine = ['Sex', 'Smoke Status', 'Chromosome', 'Hugo_Symbol', 'SNP_event', \"Consequence\", 'Exon_Number',\n",
    "                      \"Diagnosis Age\", \"TMB (nonsynonymous)\", \"Position\", \"Protein_position\", \"Codons\", \"VAR_TYPE_SX\"]"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "cancer_probabilities = {cancer_type: data_for_lift[cancer_type].mean() for cancer_type in list(data_for_lift[\"Cancer Type\"].unique())}",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "cancer_probabilities",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "    # Iterate over feature combinations\n",
    "# for num_features in range(2, 6):\n",
    "feature_combinations = list(combinations(columns_to_combine, 5))\n",
    "feature_combinations"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Precompute the mean probabilities for cancer types\n",
    "\n",
    "lifts = []\n",
    "\n",
    "for cancer_type, P_B in cancer_probabilities.items():\n",
    "    for feature in feature_combinations:\n",
    "        # Combine the selected features into a single feature\n",
    "        combined_feature = data_for_lift[list(feature)].astype(str).agg('_'.join, axis=1)\n",
    "\n",
    "        # Compute value counts for the combined feature\n",
    "        combined_counts = combined_feature.value_counts()\n",
    "        valid_features = combined_counts[combined_counts >= 100].index\n",
    "\n",
    "        if valid_features.empty:\n",
    "            continue  # Skip if no valid combined features\n",
    "\n",
    "        # Filter the combined feature to include only valid entries\n",
    "        filtered_data = combined_feature[combined_feature.isin(valid_features)]\n",
    "        P_A = filtered_data.value_counts(normalize=True)\n",
    "\n",
    "        # Compute joint probabilities for cancer type\n",
    "        joint_prob = (\n",
    "            filtered_data[data_for_lift[cancer_type] == 1]\n",
    "            .value_counts(normalize=True)\n",
    "            .reindex(P_A.index, fill_value=0)\n",
    "        )\n",
    "\n",
    "        # Calculate lift\n",
    "        lift = (joint_prob / (P_A * P_B)).round(2)\n",
    "\n",
    "        # Store results\n",
    "        lifts.append((cancer_type, feature, lift))\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "lifts = []\n",
    "# Probability of the cancer type\n",
    "for num_features in range(2, 6):\n",
    "    feature_combinations = list(combinations(columns_to_combine, num_features))\n",
    "    for cancer_type in cancer_type_dummy.columns:\n",
    "        P_B = data_for_lift[cancer_type].mean()\n",
    "        for feature in feature_combinations:\n",
    "            # Create a combined feature from three columns\n",
    "            combined_feature = data_for_lift[feature[0]].astype(str)\n",
    "\n",
    "            for f in feature[1:]:\n",
    "                combined_feature += \"_\" + data_for_lift[f].astype(str)\n",
    "            # combined_feature = \"_\".join(data_for_lift[feature].astype(str) for feature in feature_combinations)\n",
    "\n",
    "            min_count = 100\n",
    "            P_A_counts = combined_feature.value_counts()\n",
    "\n",
    "            # Filter combined features based on minimum count\n",
    "            valid_features = P_A_counts[P_A_counts >= min_count].index\n",
    "            filtered_data = combined_feature[combined_feature.isin(valid_features)]\n",
    "\n",
    "            # Probability of the combined feature\n",
    "            P_A = filtered_data.value_counts(normalize=True)\n",
    "\n",
    "            # Joint probability of the combined feature and cancer type\n",
    "            joint = (filtered_data[data_for_lift[cancer_type] == 1].value_counts(normalize=True).reindex(P_A.index, fill_value=0))\n",
    "\n",
    "            # Calculate lift\n",
    "            lift = (joint / (P_A * P_B)).round(2)  # Round lift to 2 decimal places for readability\n",
    "\n",
    "            # Append results as a tuple of the feature triplet and their associated lift values\n",
    "            lifts.append((cancer_type, feature, lift))"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Flatten the results for easy visualization\n",
    "lift_results = []\n",
    "\n",
    "for cancer_type, feature_pair, lift in lifts:\n",
    "    for feature_value, lift_value in lift.items():\n",
    "        lift_results.append({\n",
    "            'Cancer Type': cancer_type,\n",
    "            'Feature Pair': feature_pair,\n",
    "            'Feature Value': feature_value,\n",
    "            'Lift': lift_value\n",
    "        })\n",
    "\n",
    "lift_df = pd.DataFrame(lift_results)\n",
    "lift_df = lift_df.sort_values(by='Lift', ascending=False)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "lift_df.sort_values(by='Lift', ascending=False)",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "filter_triple_data = data[data[\"Smoke Status\"] == \"Nonsmoker\"]\n",
    "filter_triple_data = filter_triple_data[filter_triple_data[\"Hugo_Symbol\"] == \"TP53\"]\n",
    "filter_triple_data = filter_triple_data[filter_triple_data[\"SNP_event\"] == \"G>A\"]"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "def combine_features(data, feature_combination):\n",
    "    \"\"\"\n",
    "    Combine selected features into a single feature by joining their values.\n",
    "    \"\"\"\n",
    "    return data[list(feature_combination)].astype(str).agg('_'.join, axis=1)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "def filter_and_compute_probabilities(combined_feature, data_for_lift, cancer_type, min_count):\n",
    "    \"\"\"\n",
    "    Filter valid features, compute P(A), and joint probabilities for a cancer type.\n",
    "    \"\"\"\n",
    "    # Step 1: Compute value counts\n",
    "    combined_counts = combined_feature.value_counts()\n",
    "    valid_features = combined_counts[combined_counts >= min_count].index\n",
    "\n",
    "    # Step 2: Skip if no valid combined features\n",
    "    if valid_features.empty:\n",
    "        return None, None\n",
    "\n",
    "    # Step 3: Filter the combined feature\n",
    "    filtered_data = combined_feature[combined_feature.isin(valid_features)]\n",
    "\n",
    "    # Step 4: Compute probabilities\n",
    "    P_A = filtered_data.value_counts(normalize=True)\n",
    "    joint_prob = (\n",
    "        filtered_data[data_for_lift[cancer_type] == 1]\n",
    "        .value_counts(normalize=True)\n",
    "        .reindex(P_A.index, fill_value=0)\n",
    "    )\n",
    "\n",
    "    return P_A, joint_prob\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "def calculate_lifts_for_cancer_type(data_for_lift, cancer_type, P_B, feature_combinations, min_count):\n",
    "    \"\"\"\n",
    "    Compute lifts for a single cancer type across all feature combinations.\n",
    "    \"\"\"\n",
    "    lifts = []\n",
    "\n",
    "    for feature_combination in feature_combinations:\n",
    "        # Step 1: Combine features\n",
    "        combined_feature = combine_features(data_for_lift, feature_combination)\n",
    "\n",
    "        # Step 2: Filter and compute probabilities\n",
    "        P_A, joint_prob = filter_and_compute_probabilities(\n",
    "            combined_feature, data_for_lift, cancer_type, min_count\n",
    "        )\n",
    "\n",
    "        if P_A is None or joint_prob is None:\n",
    "            continue  # Skip if no valid combined features\n",
    "\n",
    "        # Step 3: Calculate lift\n",
    "        lift = (joint_prob / (P_A * P_B)).round(2)\n",
    "\n",
    "        # Step 4: Store result\n",
    "        lifts.append((cancer_type, feature_combination, lift))\n",
    "\n",
    "    return lifts"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "def compute_all_lifts(data_for_lift, cancer_probabilities, feature_combinations, min_count=100):\n",
    "    \"\"\"\n",
    "    Main function to compute lifts for all cancer types.\n",
    "    \"\"\"\n",
    "    all_lifts = []\n",
    "\n",
    "    for cancer_type, P_B in cancer_probabilities.items():\n",
    "        lifts = calculate_lifts_for_cancer_type(\n",
    "            data_for_lift, cancer_type, P_B, feature_combinations, min_count\n",
    "        )\n",
    "        all_lifts.extend(lifts)\n",
    "\n",
    "    return all_lifts"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "lifts = compute_all_lifts(data_for_lift, cancer_probabilities, feature_combinations, min_count=100)",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "data_dr = pd.read_csv('data_for_rules.csv')",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "data_dr.columns",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-05T13:55:55.175556Z",
     "start_time": "2025-05-05T13:55:55.040120Z"
    }
   },
   "cell_type": "code",
   "source": "narrowed = pd.read_csv(\"narrowed_cancers_data.csv\")",
   "outputs": [],
   "execution_count": 25
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "narrowed[narrowed.index == 529]",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from sklearn.impute import KNNImputer\n",
    "narrowed[['Current_Exon', 'Total_Exons']] = narrowed['Exon_Number'].str.split('/', expand=True)\n",
    "narrowed[['Current_Exon', 'Total_Exons']] = narrowed[['Current_Exon', 'Total_Exons']].astype(float)\n",
    "\n",
    "imputer = KNNImputer(n_neighbors=5)\n",
    "narrowed[[\"Current_Exon\", \"Total_Exons\"]] = imputer.fit_transform(narrowed[[\"Current_Exon\", \"Total_Exons\"]])\n",
    "# ransform(narrowed[[\"Exon_Number\"]])"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "narrowed[[\"Exon_Number\", \"Current_Exon\", \"Total_Exons\"]]",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "df = pd.read_csv(\"data_for_rules.csv\")",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "df.head()",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "df = df[df[\"Diagnosis Age\"] < 71.5]\n",
    "df = df[df[\"Diagnosis Age\"] > 47.5]\n",
    "df = df[df[\"TMB (nonsynonymous)\"] < 0.28]\n",
    "df = df[df[\"Sex\"] == \"Female\"]\n",
    "df"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "df[\"Cancer Type\"].drop_duplicates()",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "df[[\"PATIENT_ID\", \"Cancer Type\"]].drop_duplicates()[\"Cancer Type\"].value_counts()",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "df",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "raw",
   "source": "create dataframe for decision tree sentences"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "df_rules = pd.read_csv(\"data_for_rules.csv\")",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "df_rules['Exon_Number'].value_counts()",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Smoke Status - Convert to dummies\n",
    "dummy_smoking = df_rules['Smoke Status'].str.get_dummies().groupby(level=0).sum()\n",
    "\n",
    "# Hugo Symbol - Convert to dummies\n",
    "dummy_hugo_symbol = df_rules['Hugo_Symbol'].str.get_dummies().groupby(level=0).sum()\n",
    "\n",
    "# Variant Type - Convert to dummies\n",
    "dummy_Variant_Type = df_rules['Variant_Type'].str.get_dummies().groupby(level=0).sum()\n",
    "\n",
    "# SNP_event - Keep only top 100 most frequent values\n",
    "top_100 = df_rules['SNP_event'].value_counts().nlargest(100).index\n",
    "df_rules['SNP_event'] = df_rules['SNP_event'].where(df_rules['SNP_event'].isin(top_100), other=None)\n",
    "dummy_snp_event = df_rules['SNP_event'].str.get_dummies().groupby(level=0).sum()\n",
    "\n",
    "# Combine all dummy variables\n",
    "dummy_vars = pd.concat([dummy_smoking, dummy_hugo_symbol, dummy_snp_event, dummy_Variant_Type], axis=1)\n",
    "\n",
    "# Join with original DataFrame\n",
    "df_rules = df_rules.join(dummy_vars)\n",
    "\n",
    "# Drop original categorical columns\n",
    "df_rules.drop(['Smoke Status', 'SNP_event', 'Hugo_Symbol', 'Variant_Type'], axis=1, inplace=True)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "df_rules.head()",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "len(list(data_for_model[\"Event_Info\"].unique()))",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "df_rules[\"Cancer Type\"].unique()",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "df_rules['Exon_Number'] = df_rules['Exon_Number'].str.split('/').str[0].astype(int)",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "df_rules.drop('Codons', axis=1, inplace=True)",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "df_rules.head(10)",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "df_rules.to_csv(\"data_for_decision.csv\", index=False)",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "df = data_for_model.copy()\n",
    "df.drop(['Site1_Hugo_Symbol', 'Site2_Hugo_Symbol', 'Event_Info'], axis=1, inplace=True)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "data_for_model.dropna(inplace=True)\n",
    "data_for_model['Exon_Number'] = data_for_model['Exon_Number'].str.split('/').str[0].astype(int)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Step 1: Prepare your categorical features\n",
    "categorical_features = ['Sex', 'VAR_TYPE_SX', 'Smoke Status', 'Hugo_Symbol', 'Variant_Type', 'SNP_event', 'Consequence', 'Chromosome']\n",
    "\n",
    "# Step 2: Handle the special case of Codons\n",
    "def prepare_data(df, columns):\n",
    "    # For high-cardinality features like Codons, we can group them\n",
    "    # Example: Group by first letter of codon or by some domain knowledge\n",
    "\n",
    "    # Method 1: Keep only the most frequent codons and group others\n",
    "    for column in columns:\n",
    "        top_codons = df[column].value_counts().nlargest(100).index.tolist()\n",
    "        df[f'{column}_grouped'] = df[column].apply(lambda x: x if x in top_codons else 'Other')\n",
    "\n",
    "    # OR Method 2: Group by first nucleotide\n",
    "    # df['Codons_grouped'] = df['Codons'].apply(lambda x: x[0] + '_codons' if isinstance(x, str) else 'Unknown')\n",
    "\n",
    "    return df\n",
    "\n",
    "# Step 3: Encoding categorical features\n",
    "def encode_features(df, categorical_cols):\n",
    "    label_encoders = {}\n",
    "\n",
    "    # Store original values for interpretation\n",
    "    feature_values = {}\n",
    "\n",
    "    for col in categorical_cols + ['SNP_event_grouped']:\n",
    "        if col in df.columns:\n",
    "            le = LabelEncoder()\n",
    "            df[col + '_encoded'] = le.fit_transform(df[col])\n",
    "\n",
    "            # Store mapping for interpretation\n",
    "            label_encoders[col] = le\n",
    "            feature_values[col] = dict(zip(le.transform(le.classes_), le.classes_))\n",
    "\n",
    "    return df, label_encoders, feature_values\n",
    "\n",
    "def extract_rules(clf, feature_names, class_names, feature_values):\n",
    "    tree_ = clf.tree_\n",
    "\n",
    "    feature_name = [\n",
    "        feature_names[i] if i != tree._tree.TREE_UNDEFINED else \"undefined!\"\n",
    "        for i in tree_.feature\n",
    "    ]\n",
    "\n",
    "    paths = []\n",
    "\n",
    "    def recurse(node, path, paths):\n",
    "        if tree_.feature[node] != tree._tree.TREE_UNDEFINED:\n",
    "            name = feature_name[node]\n",
    "            threshold = tree_.threshold[node]\n",
    "\n",
    "            # Special handling for chromosome dummy variables\n",
    "            if name.startswith('chr_'):\n",
    "                # Extract chromosome number\n",
    "                chr_num = name.split('_')[1]\n",
    "\n",
    "                # For dummy variables, typically threshold is 0.5\n",
    "                if threshold <= 0.5:\n",
    "                    # chr_X ≤ 0.5 means the mutation is NOT on this chromosome\n",
    "                    path.append((name, \"chromosome\", {\"excluded\": chr_num}, f\"Chromosome is not {chr_num}\"))\n",
    "                    recurse(tree_.children_left[node], path, paths)\n",
    "                    path.pop()\n",
    "\n",
    "                    # chr_X > 0.5 means the mutation IS on this chromosome\n",
    "                    path.append((name, \"chromosome\", {\"included\": chr_num}, f\"Chromosome is {chr_num}\"))\n",
    "                    recurse(tree_.children_right[node], path, paths)\n",
    "                    path.pop()\n",
    "            # Handle categorical features\n",
    "            elif name.endswith('_encoded'):\n",
    "                original_name = name.replace('_encoded', '')\n",
    "\n",
    "                if original_name in feature_values:\n",
    "                    # Handle categorical feature\n",
    "                    left_values = [feature_values[original_name][i] for i in range(len(feature_values[original_name]))\n",
    "                                  if i <= threshold]\n",
    "                    right_values = [feature_values[original_name][i] for i in range(len(feature_values[original_name]))\n",
    "                                   if i > threshold]\n",
    "\n",
    "                    # Store as tuples: (feature_name, \"categorical\", values_list, readable_condition)\n",
    "                    if len(left_values) <= 3:\n",
    "                        left_condition = f\"{original_name} is {' or '.join(map(str, left_values))}\"\n",
    "                    else:\n",
    "                        left_condition = f\"{original_name} is in a group of {len(left_values)} values\"\n",
    "\n",
    "                   # Handle special case for Codons\n",
    "                    if original_name == 'SNP_event_grouped' and 'Other' in left_values:\n",
    "                        left_condition = f\"SNP event is among the less common types\"\n",
    "                    if original_name == 'SNP_event_grouped' and 'Other' in right_values:\n",
    "                        right_condition = f\"SNP event is among the less common types\"\n",
    "\n",
    "                    path.append((original_name, \"categorical\", set(left_values), left_condition))\n",
    "                    recurse(tree_.children_left[node], path, paths)\n",
    "                    path.pop()\n",
    "\n",
    "                    if len(right_values) <= 3:\n",
    "                        right_condition = f\"{original_name} is {' or '.join(map(str, right_values))}\"\n",
    "                    else:\n",
    "                        right_condition = f\"{original_name} is in a group of {len(right_values)} values\"\n",
    "\n",
    "                    path.append((original_name, \"categorical\", set(right_values), right_condition))\n",
    "                    recurse(tree_.children_right[node], path, paths)\n",
    "                    path.pop()\n",
    "                else:\n",
    "                    # Standard case for encoded features without mapping\n",
    "                    path.append((original_name, \"categorical\", {f\"≤ category {threshold:.0f}\"},\n",
    "                                f\"{original_name} ≤ category {threshold:.0f}\"))\n",
    "                    recurse(tree_.children_left[node], path, paths)\n",
    "                    path.pop()\n",
    "\n",
    "                    path.append((original_name, \"categorical\", {f\"> category {threshold:.0f}\"},\n",
    "                                f\"{original_name} > category {threshold:.0f}\"))\n",
    "                    recurse(tree_.children_right[node], path, paths)\n",
    "                    path.pop()\n",
    "            else:\n",
    "                # Numerical features - ensure we use consistent 4-tuple format\n",
    "                left_condition = f\"{name} ≤ {threshold:.2f}\"\n",
    "                path.append((name, \"numerical\", {\"min\": float(\"-inf\"), \"max\": threshold}, left_condition))\n",
    "                recurse(tree_.children_left[node], path, paths)\n",
    "                path.pop()\n",
    "\n",
    "                right_condition = f\"{name} > {threshold:.2f}\"\n",
    "                path.append((name, \"numerical\", {\"min\": threshold, \"max\": float(\"inf\")}, right_condition))\n",
    "                recurse(tree_.children_right[node], path, paths)\n",
    "                path.pop()\n",
    "        else:\n",
    "            class_idx = np.argmax(tree_.value[node][0])\n",
    "            paths.append((path.copy(), class_names[class_idx]))\n",
    "\n",
    "    recurse(0, [], paths)\n",
    "\n",
    "    # Generate human-readable sentences with consolidated features\n",
    "    rules = []\n",
    "    for path, outcome in paths:\n",
    "        if path:\n",
    "            # Group by feature name\n",
    "            feature_groups = {}\n",
    "            for condition in path:\n",
    "                feature, cond_type, value_info, readable = condition  # Now this should always work\n",
    "                if feature not in feature_groups:\n",
    "                    feature_groups[feature] = []\n",
    "                feature_groups[feature].append((cond_type, value_info, readable))\n",
    "\n",
    "            # Process chromosome features\n",
    "            chromosomes_included = []\n",
    "            chromosomes_excluded = []\n",
    "            other_feature_groups = {}\n",
    "\n",
    "            for feature, conditions in feature_groups.items():\n",
    "                if any(c[0] == \"chromosome\" for c in conditions):\n",
    "                    for cond_type, value_info, _ in conditions:\n",
    "                        if \"included\" in value_info:\n",
    "                            chromosomes_included.append(value_info[\"included\"])\n",
    "                        if \"excluded\" in value_info:\n",
    "                            chromosomes_excluded.append(value_info[\"excluded\"])\n",
    "                else:\n",
    "                    other_feature_groups[feature] = conditions\n",
    "\n",
    "            # Create consolidated conditions\n",
    "            consolidated_conditions = []\n",
    "\n",
    "            # Add chromosome conditions\n",
    "            if chromosomes_included:\n",
    "                if len(chromosomes_included) == 1:\n",
    "                    consolidated_conditions.append(f\"Chromosome is {chromosomes_included[0]}\")\n",
    "                else:\n",
    "                    consolidated_conditions.append(f\"Chromosome is one of {', '.join(chromosomes_included)}\")\n",
    "\n",
    "            if chromosomes_excluded:\n",
    "                if len(chromosomes_excluded) <= 3:\n",
    "                    consolidated_conditions.append(f\"Chromosome is not {', '.join(chromosomes_excluded)}\")\n",
    "\n",
    "            # Process other features\n",
    "            for feature, conditions in other_feature_groups.items():\n",
    "                if all(c[0] == \"numerical\" for c in conditions):\n",
    "                    # For numerical features\n",
    "                    min_val = float(\"-inf\")\n",
    "                    max_val = float(\"inf\")\n",
    "\n",
    "                    for _, value_info, _ in conditions:\n",
    "                        min_val = max(min_val, value_info.get(\"min\", float(\"-inf\")))\n",
    "                        max_val = min(max_val, value_info.get(\"max\", float(\"inf\")))\n",
    "\n",
    "                    if min_val > float(\"-inf\") and max_val < float(\"inf\"):\n",
    "                        consolidated_conditions.append(f\"{feature} is between {min_val:.2f} and {max_val:.2f}\")\n",
    "                    elif min_val > float(\"-inf\"):\n",
    "                        consolidated_conditions.append(f\"{feature} > {min_val:.2f}\")\n",
    "                    elif max_val < float(\"inf\"):\n",
    "                        consolidated_conditions.append(f\"{feature} ≤ {max_val:.2f}\")\n",
    "\n",
    "                elif all(c[0] == \"categorical\" for c in conditions):\n",
    "                    # For categorical features - find intersection of values\n",
    "                    value_sets = [c[1] for c in conditions]\n",
    "\n",
    "                    # Find intersection of all sets\n",
    "                    common_values = set.intersection(*value_sets) if value_sets else set()\n",
    "\n",
    "                    # If intersection is non-empty, it's the stricter condition\n",
    "                    if common_values:\n",
    "                        if len(common_values) <= 3:\n",
    "                            consolidated_conditions.append(f\"{feature} is {' or '.join(map(str, common_values))}\")\n",
    "                        else:\n",
    "                            consolidated_conditions.append(f\"{feature} is in a group of {len(common_values)} values\")\n",
    "                    else:\n",
    "                        # If no intersection (shouldn't happen in a valid tree), use original conditions\n",
    "                        for _, _, readable in conditions:\n",
    "                            consolidated_conditions.append(readable)\n",
    "\n",
    "            # Create the final rule\n",
    "            rule = \"If \" + \" AND \".join(consolidated_conditions) + f\", THEN cancer type is {outcome}\"\n",
    "            rules.append(rule)\n",
    "\n",
    "    return rules"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "df = df_rules.copy()\n",
    "df.head()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# d = prepare_data(df)\n",
    "df, label_encoders, feature_values = encode_features(df, ['Sex', 'VAR_TYPE_SX'])\n",
    "df.head()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "features_to_drop = ['Cancer Type', 'Cancer Type Detailed', 'Tumor Stage', 'Sample Type', 'Sex', 'VAR_TYPE_SX']\n",
    "                    # 'Smoke Status', 'Hugo_Symbol', 'Variant_Type', 'SNP_event', 'SNP_event_grouped', 'Codons', 'Consequence', 'Chromosome']\n",
    "y = df['Cancer Type']\n",
    "X = df.drop(features_to_drop, axis=1)\n",
    "X_train, X_test, y_train, y_test, X_test_with_id = stratified_split_by_patient(X, y)\n",
    "feature_names = list(X_train.columns)\n",
    "class_names = list(df['Cancer Type'].unique())\n",
    "clf = tree.DecisionTreeClassifier(random_state=39)#, min_samples_leaf=10)#, max_depth=1000)\n",
    "clf.fit(X_train, y_train)\n",
    "sentences = extract_rules(clf, feature_names, class_names, feature_values)\n",
    "\n",
    "# for sentence in sentences:\n",
    "#     print(sentence)\n",
    "sentences"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "feature_names",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "data_for_model.head()",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "df = prepare_data(data_for_model, ['SNP_event', 'Codons',])",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "df = pd.read_csv('models_hypotheses/combined_hypotheses.csv')\n",
    "df['plausibility'] = None\n",
    "df['novelty'] = None\n",
    "df['comments'] = None"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "df[df['rank'] <= 10].to_excel('models_hypotheses/hypotheses_for_professional_evaluation.xlsx', index=False)",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "df['Tumor Stage'].value_counts()",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "data = df[df['Tumor Stage'] in ['III', 'IV', 'II', 'I']]",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "def get_latest_csv(directory):\n",
    "    list_of_files = glob.glob(os.path.join(directory, '*.csv'))\n",
    "    if not list_of_files:\n",
    "        raise FileNotFoundError(\"No CSV files found in the directory.\")\n",
    "    latest_file = max(list_of_files, key=os.path.getmtime)\n",
    "    print(f\"Using file: {latest_file}\")\n",
    "    return latest_file"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "latest_csv_path = get_latest_csv(\"llm_results\")\n",
    "df = pd.read_csv(latest_csv_path)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "# df = pd.read_csv('llm_results/evaluations_20250409_151350.csv')",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "combine_hyp = pd.read_csv('models_hypotheses/combined_hypotheses.csv')\n",
    "combine_hyp.rename(columns={'hypo_id': 'hypothesis_id'}, inplace=True)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Make sure both columns are the same type (e.g., convert both to string or both to int)\n",
    "df['hypothesis_id'] = df['hypothesis_id'].astype(str)\n",
    "combine_hyp['hypothesis_id'] = combine_hyp['hypothesis_id'].astype(str)\n",
    "\n",
    "# Now you can safely join\n",
    "merged_df = df.join(combine_hyp.set_index('hypothesis_id'), on='hypothesis_id')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "merged_df.sort_values(by=['novelty', 'plausibility'], ascending=[False, False], inplace=True)",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "data = merged_df[merged_df['novelty'] >= 6]\n",
    "data = data[data['plausibility'] >= 6]\n",
    "data = data[(data['novelty'] > 6) | (data['plausibility'] > 6)]\n",
    "# data['novelty'] = None\n",
    "# data['plausibility'] = None\n",
    "data.drop(columns=['timestamp'], inplace=True)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "merged_df[(merged_df['novelty'] == merged_df['plausibility'])]\n",
    "# merged_df[merged_df[\"model\"] == \"openai:o3-mini\"]"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "data.to_excel('models_hypotheses/hypotheses_for_professional2.xlsx', index=False)",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from catboost import CatBoostClassifier, Pool\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from sklearn.model_selection import train_test_split, StratifiedKFold\n",
    "from sklearn.metrics import classification_report, confusion_matrix, accuracy_score\n",
    "import shap\n",
    "from imblearn.over_sampling import SMOTE\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.decomposition import PCA\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "def load_data(filepath):\n",
    "    features_to_drop = ['Cancer Type Detailed', 'Tumor Stage', 'Sample Type', \"Site2_Hugo_Symbol\", \"Site1_Hugo_Symbol\", \"Event_Info\"]\n",
    "    label = 'Cancer Type'\n",
    "    data = pd.read_csv(filepath)\n",
    "    cancer_types = data[\"Cancer Type\"].unique()\n",
    "    # mapping = {}\n",
    "    # # Convert object columns to categorical\n",
    "    # object_columns = data.select_dtypes(include=['object', 'bool']).columns\n",
    "    # for col in object_columns:\n",
    "    #     mapping[col] = dict(enumerate(data[col].astype('category').cat.categories))\n",
    "    # data[object_columns] = data[object_columns].astype('category')\n",
    "    #\n",
    "    # # Encode categorical columns using cat.codes\n",
    "    # for col in data.select_dtypes(include='category').columns:\n",
    "    #     data[col] = data[col].cat.codes\n",
    "\n",
    "    # Separate features and labels\n",
    "    data.drop(features_to_drop, axis=1, inplace=True)\n",
    "    data.dropna(inplace=True)\n",
    "    X = data.drop(label, axis=1)\n",
    "    y, uniques = pd.factorize(data['Cancer Type'])\n",
    "    # label_dict = {cancer: idx for idx, cancer in enumerate(cancer_types)}\n",
    "    # X.replace(-1, np.nan, inplace=True)\n",
    "    return X, y#, label_dict, mapping\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "def stratified_split_by_patient(X, y, train_ratio=0.7, test_ratio=0.3):\n",
    "    \"\"\"\n",
    "    Split data into training and testing sets with stratification by PATIENT_ID.\n",
    "    \"\"\"\n",
    "    # Ensure the ratios sum to 1\n",
    "    assert train_ratio + test_ratio == 1, \"Ratios must sum to 1.\"\n",
    "\n",
    "    # Get unique patient IDs\n",
    "    unique_ids = X['PATIENT_ID'].unique()\n",
    "\n",
    "    # Map PATIENT_ID to a corresponding target value (first occurrence)\n",
    "    patient_labels = dict(zip(X['PATIENT_ID'], y))\n",
    "    unique_patient_labels = [patient_labels[pid] for pid in unique_ids]\n",
    "\n",
    "    # Initial split: train+val and test\n",
    "    train_ids, test_ids = train_test_split(\n",
    "        unique_ids,\n",
    "        test_size=test_ratio,\n",
    "        stratify=unique_patient_labels,\n",
    "        random_state=42\n",
    "    )\n",
    "\n",
    "    # Split data into subsets\n",
    "    X_train = X[X['PATIENT_ID'].isin(train_ids)].drop(columns=['PATIENT_ID'])\n",
    "    X_test = X[X['PATIENT_ID'].isin(test_ids)].drop(columns=['PATIENT_ID'])\n",
    "    X_test_with_id = X[X['PATIENT_ID'].isin(test_ids)]  # Keep validation set with PATIENT_ID for patient-level analysis\n",
    "\n",
    "    y_train = y[X['PATIENT_ID'].isin(train_ids)]\n",
    "    # y_val = y[X['PATIENT_ID'].isin(val_ids)]\n",
    "    y_test = y[X['PATIENT_ID'].isin(test_ids)]\n",
    "\n",
    "    return X_train, X_test, y_train, y_test, X_test_with_id"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "def analyze_feature_combinations_for_cancer(model, X, y, cancer_type, cancer_names,\n",
    "                                           top_n=5, cat_features=None,\n",
    "                                           interaction_depth=2):\n",
    "    \"\"\"\n",
    "    Analyze feature combinations specific to a cancer type\n",
    "\n",
    "    Parameters:\n",
    "    -----------\n",
    "    model : CatBoostClassifier\n",
    "        Trained model\n",
    "    X : DataFrame\n",
    "        Feature data\n",
    "    y : Series\n",
    "        Target labels\n",
    "    cancer_type : int\n",
    "        The specific cancer type to analyze\n",
    "    cancer_names : list\n",
    "        List of cancer type names corresponding to encoded values\n",
    "    top_n : int\n",
    "        Number of top combinations to return\n",
    "    cat_features : list\n",
    "        List of categorical feature names\n",
    "    interaction_depth : int\n",
    "        Max number of features to consider in combinations\n",
    "    \"\"\"\n",
    "    import numpy as np\n",
    "    import pandas as pd\n",
    "    import matplotlib.pyplot as plt\n",
    "    import seaborn as sns\n",
    "    import shap\n",
    "\n",
    "    # Get actual cancer name\n",
    "    cancer_name = cancer_names[cancer_type]\n",
    "\n",
    "    # Get samples for this cancer type\n",
    "    cancer_indices = np.where(y == cancer_type)[0]\n",
    "    X_cancer = X.iloc[cancer_indices]\n",
    "\n",
    "    # Get samples for other cancer types\n",
    "    other_indices = np.where(y != cancer_type)[0]\n",
    "    X_other = X.iloc[other_indices]\n",
    "\n",
    "    print(f\"\\n===== ANALYZING CANCER TYPE: {cancer_name} =====\")\n",
    "    print(f\"Number of samples: {len(X_cancer)}\")\n",
    "\n",
    "    # Get feature importance for this specific cancer type\n",
    "    # Use SHAP values for better feature importance\n",
    "    try:\n",
    "        print(\"\\nCalculating SHAP values for feature importance...\")\n",
    "        explainer = shap.TreeExplainer(model)\n",
    "\n",
    "        # Get class-specific SHAP values\n",
    "        # Sample for efficiency if dataset is large\n",
    "        sample_size = min(len(X), 1000)\n",
    "        X_sample = X.sample(sample_size, random_state=42)\n",
    "        shap_values = explainer.shap_values(X_sample)\n",
    "\n",
    "        # Get SHAP values for this cancer type\n",
    "        cancer_shap = shap_values[cancer_type]\n",
    "\n",
    "        # Calculate mean absolute SHAP value for each feature\n",
    "        feature_importance = np.abs(cancer_shap).mean(0)\n",
    "        feature_names = X.columns\n",
    "\n",
    "        # Sort features by importance\n",
    "        sorted_idx = np.argsort(-feature_importance)\n",
    "\n",
    "        # Plot SHAP values\n",
    "        plt.figure(figsize=(12, 10))\n",
    "        shap.summary_plot(cancer_shap, X_sample, plot_type=\"bar\", show=False)\n",
    "        plt.title(f'SHAP Feature Importance for {cancer_name}')\n",
    "        plt.tight_layout()\n",
    "        plt.show()\n",
    "\n",
    "        # List top important features\n",
    "        print(f\"\\nTop features for {cancer_name}:\")\n",
    "        for i in range(min(10, len(sorted_idx))):\n",
    "            idx = sorted_idx[i]\n",
    "            print(f\"{feature_names[idx]}: {feature_importance[idx]:.4f}\")\n",
    "\n",
    "        # Get top features for interaction analysis\n",
    "        top_features = [feature_names[i] for i in sorted_idx[:15]]  # Use top 15 features\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"SHAP analysis error: {e}\")\n",
    "        print(\"Falling back to CatBoost feature importance\")\n",
    "\n",
    "        # Use CatBoost feature importance\n",
    "        feature_importance = model.get_feature_importance()\n",
    "        feature_names = X.columns\n",
    "        sorted_idx = np.argsort(-feature_importance)\n",
    "\n",
    "        # Plot feature importance\n",
    "        plt.figure(figsize=(12, 10))\n",
    "        plt.barh(range(min(15, len(sorted_idx))),\n",
    "                feature_importance[sorted_idx[:15]])\n",
    "        plt.yticks(range(min(15, len(sorted_idx))),\n",
    "                  [feature_names[i] for i in sorted_idx[:15]])\n",
    "        plt.title(f'Feature Importance for {cancer_name}')\n",
    "        plt.tight_layout()\n",
    "        plt.show()\n",
    "\n",
    "        # Get top features for interaction analysis\n",
    "        top_features = [feature_names[i] for i in sorted_idx[:15]]  # Use top 15 features\n",
    "\n",
    "    # Analyze feature distributions for this cancer type vs others\n",
    "    print(\"\\nAnalyzing feature distributions...\")\n",
    "\n",
    "    # Select features for analysis (mix of top important and categorical)\n",
    "    analysis_features = top_features.copy()\n",
    "    if cat_features:\n",
    "        for cf in cat_features:\n",
    "            if cf not in analysis_features:\n",
    "                analysis_features.append(cf)\n",
    "\n",
    "    # Keep unique features only\n",
    "    analysis_features = list(set(analysis_features))\n",
    "\n",
    "    # Analyze individual feature distributions\n",
    "    for feature in analysis_features[:10]:  # Limit to top 10 for clarity\n",
    "        if feature in cat_features if cat_features else []:\n",
    "            # Categorical feature analysis\n",
    "            cancer_dist = X_cancer[feature].value_counts(normalize=True)\n",
    "            other_dist = X_other[feature].value_counts(normalize=True)\n",
    "\n",
    "            # Compute lift ratio (how much more likely in this cancer type)\n",
    "            lift = pd.DataFrame({\n",
    "                'Cancer': cancer_dist,\n",
    "                'Other': other_dist\n",
    "            }).fillna(0)\n",
    "            lift['Lift'] = lift['Cancer'] / lift['Other'].replace(0, 0.001)\n",
    "            lift = lift.sort_values('Lift', ascending=False)\n",
    "\n",
    "            print(f\"\\nFeature: {feature}\")\n",
    "            print(\"Top values by lift ratio:\")\n",
    "            print(lift[['Cancer', 'Other', 'Lift']].head(3))\n",
    "\n",
    "            # Plot distribution comparison\n",
    "            plt.figure(figsize=(12, 6))\n",
    "\n",
    "            # Get top categories by frequency\n",
    "            top_cats = set(cancer_dist.nlargest(5).index) | set(other_dist.nlargest(5).index)\n",
    "\n",
    "            # Filter both distributions to these categories\n",
    "            plot_data = pd.DataFrame({\n",
    "                f'{cancer_name}': cancer_dist.reindex(top_cats).fillna(0),\n",
    "                'Other Cancer Types': other_dist.reindex(top_cats).fillna(0)\n",
    "            })\n",
    "\n",
    "            plot_data.plot(kind='bar', ax=plt.gca())\n",
    "            plt.title(f'Distribution of {feature} ({cancer_name} vs Others)')\n",
    "            plt.ylabel('Frequency')\n",
    "            plt.xticks(rotation=45)\n",
    "            plt.tight_layout()\n",
    "            plt.show()\n",
    "        else:\n",
    "            # Numerical feature analysis - Fix for TypeError\n",
    "            plt.figure(figsize=(12, 6))\n",
    "\n",
    "            # Check if feature is numeric\n",
    "            if pd.api.types.is_numeric_dtype(X[feature]):\n",
    "                # KDE plot for numeric data\n",
    "                sns.kdeplot(X_cancer[feature].astype(float), label=f'{cancer_name}')\n",
    "                sns.kdeplot(X_other[feature].astype(float), label='Other Cancer Types')\n",
    "                plt.title(f'Distribution of {feature} ({cancer_name} vs Others)')\n",
    "                plt.xlabel(feature)\n",
    "                plt.ylabel('Density')\n",
    "                plt.legend()\n",
    "            else:\n",
    "                # For non-numeric data, use countplot instead of kdeplot\n",
    "                plot_data = pd.DataFrame({\n",
    "                    'value': pd.concat([X_cancer[feature], X_other[feature]]),\n",
    "                    'group': (['This Cancer'] * len(X_cancer)) + (['Other Cancers'] * len(X_other))\n",
    "                })\n",
    "                sns.countplot(x='value', hue='group', data=plot_data)\n",
    "                plt.title(f'Distribution of {feature} ({cancer_name} vs Others)')\n",
    "                plt.xlabel(feature)\n",
    "                plt.ylabel('Count')\n",
    "                plt.xticks(rotation=45)\n",
    "\n",
    "            plt.tight_layout()\n",
    "            plt.show()\n",
    "\n",
    "            # Calculate statistics for numeric features\n",
    "            if pd.api.types.is_numeric_dtype(X[feature]):\n",
    "                cancer_mean = X_cancer[feature].mean()\n",
    "                other_mean = X_other[feature].mean()\n",
    "                mean_diff_pct = ((cancer_mean - other_mean) / other_mean * 100\n",
    "                                if other_mean != 0 else float('inf'))\n",
    "\n",
    "                print(f\"\\nFeature: {feature}\")\n",
    "                print(f\"Mean for {cancer_name}: {cancer_mean:.4f}\")\n",
    "                print(f\"Mean for Other Cancer Types: {other_mean:.4f}\")\n",
    "                print(f\"Difference: {mean_diff_pct:.2f}%\")\n",
    "\n",
    "    # Feature interaction analysis\n",
    "    if cat_features and len(cat_features) >= 2:\n",
    "        print(\"\\nAnalyzing categorical feature interactions...\")\n",
    "\n",
    "        # Select categorical features for analysis\n",
    "        cat_features_for_analysis = [f for f in analysis_features if f in cat_features]\n",
    "        cat_features_for_analysis = cat_features_for_analysis[:min(len(cat_features_for_analysis), 5)]\n",
    "\n",
    "        if len(cat_features_for_analysis) >= 2:\n",
    "            feature_combinations = []\n",
    "\n",
    "            # Analyze pairwise combinations\n",
    "            for i, feat1 in enumerate(cat_features_for_analysis):\n",
    "                for feat2 in cat_features_for_analysis[i+1:]:\n",
    "                    # Get value counts for both features in cancer subset\n",
    "                    combo_cancer = X_cancer.groupby([feat1, feat2]).size().reset_index()\n",
    "                    combo_cancer.columns = [feat1, feat2, 'cancer_count']\n",
    "                    combo_cancer['cancer_pct'] = combo_cancer['cancer_count'] / len(X_cancer) * 100\n",
    "\n",
    "                    # Get value counts for both features in other subset\n",
    "                    combo_other = X_other.groupby([feat1, feat2]).size().reset_index()\n",
    "                    combo_other.columns = [feat1, feat2, 'other_count']\n",
    "                    combo_other['other_pct'] = combo_other['other_count'] / len(X_other) * 100\n",
    "\n",
    "                    # Merge and calculate lift\n",
    "                    combo_merged = pd.merge(combo_cancer, combo_other, on=[feat1, feat2], how='left')\n",
    "                    combo_merged.fillna({'other_count': 0, 'other_pct': 0.001}, inplace=True)\n",
    "                    combo_merged['lift'] = combo_merged['cancer_pct'] / combo_merged['other_pct']\n",
    "\n",
    "                    # Filter to significant combinations (with enough samples)\n",
    "                    combo_merged = combo_merged[combo_merged['cancer_count'] >= 5]\n",
    "\n",
    "                    # Sort by lift\n",
    "                    combo_merged.sort_values('lift', ascending=False, inplace=True)\n",
    "\n",
    "                    # Add top combinations to results\n",
    "                    for _, row in combo_merged.head(3).iterrows():\n",
    "                        feature_combinations.append({\n",
    "                            'Feature1': feat1,\n",
    "                            'Value1': row[feat1],\n",
    "                            'Feature2': feat2,\n",
    "                            'Value2': row[feat2],\n",
    "                            'Cancer_Count': row['cancer_count'],\n",
    "                            'Cancer_Pct': row['cancer_pct'],\n",
    "                            'Other_Pct': row['other_pct'],\n",
    "                            'Lift': row['lift']\n",
    "                        })\n",
    "\n",
    "            # Sort all combinations by lift and display top results\n",
    "            if feature_combinations:\n",
    "                combinations_df = pd.DataFrame(feature_combinations)\n",
    "                combinations_df.sort_values('Lift', ascending=False, inplace=True)\n",
    "\n",
    "                print(f\"\\nTop feature combinations for {cancer_name}:\")\n",
    "                pd.set_option('display.max_colwidth', 30)\n",
    "                print(combinations_df.head(top_n))\n",
    "\n",
    "                # Plot top combinations\n",
    "                plt.figure(figsize=(14, 8))\n",
    "                bars = plt.barh(range(len(combinations_df.head(top_n))),\n",
    "                                combinations_df.head(top_n)['Lift'],\n",
    "                                color='skyblue')\n",
    "                plt.yticks(range(len(combinations_df.head(top_n))),\n",
    "                          [f\"{row['Feature1']}={row['Value1']}, {row['Feature2']}={row['Value2']}\"\n",
    "                           for _, row in combinations_df.head(top_n).iterrows()])\n",
    "                plt.xlabel('Lift (How much more common in this cancer type)')\n",
    "                plt.title(f'Top Feature Combinations for {cancer_name}')\n",
    "\n",
    "                # Add value labels\n",
    "                for i, bar in enumerate(bars):\n",
    "                    plt.text(bar.get_width() + 0.2,\n",
    "                            bar.get_y() + bar.get_height()/2,\n",
    "                            f\"{combinations_df.iloc[i]['Lift']:.2f}x\",\n",
    "                            va='center')\n",
    "\n",
    "                plt.tight_layout()\n",
    "                plt.show()\n",
    "            else:\n",
    "                print(\"No significant feature combinations found.\")\n",
    "        else:\n",
    "            print(\"Not enough categorical features for combination analysis.\")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "def main():\n",
    "    # Import necessary libraries\n",
    "    import pandas as pd\n",
    "    import numpy as np\n",
    "    from catboost import CatBoostClassifier, Pool\n",
    "    from sklearn.metrics import classification_report\n",
    "\n",
    "    # Load your data here\n",
    "    features_to_drop = ['Cancer Type Detailed', 'Tumor Stage', 'Sample Type', \"Site2_Hugo_Symbol\", \"Site1_Hugo_Symbol\", \"Event_Info\"]\n",
    "    label = 'Cancer Type'\n",
    "    data = pd.read_csv(\"narrowed_cancers_data.csv\")\n",
    "    data.drop(features_to_drop, axis=1, inplace=True)\n",
    "    data.dropna(inplace=True)\n",
    "    X = data.drop(label, axis=1)\n",
    "\n",
    "    # Get cancer names and encoded values\n",
    "    y, cancer_names = pd.factorize(data['Cancer Type'])\n",
    "    cancer_names = cancer_names.tolist()  # Convert to list for indexing\n",
    "\n",
    "    # Split data\n",
    "    X_train, X_test, y_train, y_test, X_test_with_id = stratified_split_by_patient(X, y)\n",
    "    categorical_columns = X_train.select_dtypes(include=['object']).columns.tolist()\n",
    "\n",
    "    # Create CatBoost Pool with categorical features identified\n",
    "    train_pool = Pool(X_train, y_train, cat_features=categorical_columns)\n",
    "    test_pool = Pool(X_test, y_test, cat_features=categorical_columns)\n",
    "\n",
    "    # Initialize and train CatBoost model\n",
    "    model = CatBoostClassifier(\n",
    "        iterations=500,\n",
    "        learning_rate=0.05,\n",
    "        depth=6,\n",
    "        loss_function='MultiClass',\n",
    "        random_seed=42,\n",
    "        verbose=100\n",
    "    )\n",
    "\n",
    "    model.fit(train_pool)\n",
    "\n",
    "    # Evaluate model performance\n",
    "    print(\"\\nModel evaluation:\")\n",
    "    predictions = model.predict(X_test)\n",
    "    pred_proba = model.predict_proba(X_test)\n",
    "\n",
    "    print(\"\\nClassification Report:\")\n",
    "    print(classification_report(y_test, predictions, target_names=cancer_names))\n",
    "\n",
    "    cat_features = X_train.select_dtypes(include=['object', 'category']).columns.tolist()\n",
    "\n",
    "    # ====== FEATURE COMBINATION ANALYSIS FOR EACH CANCER TYPE ======\n",
    "    print(\"Analyzing feature combinations for each cancer type...\")\n",
    "    unique_cancer_types = np.unique(y)\n",
    "\n",
    "    for cancer_type in unique_cancer_types:\n",
    "        analyze_feature_combinations_for_cancer(\n",
    "            model, X, y, cancer_type, cancer_names,\n",
    "            top_n=5, cat_features=cat_features\n",
    "        )\n",
    "\n",
    "    print(\"Analysis complete!\")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "if __name__ == \"__main__\":\n",
    "    main()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "X, y = load_data(\"narrowed_cancers_data.csv\")\n",
    "X_train, X_test, y_train, y_test, X_test_with_id = stratified_split_by_patient(X, y)\n",
    "categorical_columns = X_train.select_dtypes(include=['object']).columns.tolist()\n",
    "print(f\"\\nCategorical columns: {categorical_columns}\")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "\n",
    "# Create CatBoost Pool with categorical features identified\n",
    "train_pool = Pool(X_train, y_train, cat_features=categorical_columns)\n",
    "test_pool = Pool(X_test, y_test, cat_features=categorical_columns)\n",
    "\n",
    "# Initialize and train CatBoost model\n",
    "model = CatBoostClassifier(\n",
    "    iterations=500,\n",
    "    learning_rate=0.05,\n",
    "    depth=6,\n",
    "    loss_function='MultiClass',\n",
    "    random_seed=42,\n",
    "    verbose=100\n",
    ")\n",
    "\n",
    "model.fit(train_pool)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "\n",
    "# Evaluate model performance\n",
    "print(\"\\nModel evaluation:\")\n",
    "predictions = model.predict(X_test)\n",
    "pred_proba = model.predict_proba(X_test)\n",
    "\n",
    "print(\"\\nClassification Report:\")\n",
    "print(classification_report(y_test, predictions))"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "",
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
